{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fd3911ad-a065-46f7-bfaf-05ad3c7c6bdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pdb\n",
    "import numpy as np\n",
    "import itertools\n",
    "from snakemake.io import expand\n",
    "import yaml\n",
    "import os\n",
    "import sys\n",
    "import pathlib\n",
    "import pyranges as pr\n",
    "import cerberus \n",
    "\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from utils import *\n",
    "from sm_utils import *\n",
    "from humanized_utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "30cabd11-d61c-44e1-a2e4-24b61b728b13",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = 'config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6673d4e-59d3-4838-a670-bdd719c1bc3e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5e1e226c-042b-4715-912f-39430f78e9af",
   "metadata": {},
   "source": [
    "## 231108 merge pseudochrom gtfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4bb063ee-8b0c-4cc2-802d-12bf4de4c139",
   "metadata": {},
   "outputs": [],
   "source": [
    "# config_tsv = '231027_config.tsv'\n",
    "# config_tsv = 'test_pseudochrom_parsing_config.tsv'\n",
    "# config_tsv = '231102_config_ad003.tsv'\n",
    "config_tsv = '231106_config_ad008_ad003.tsv'\n",
    "p_meta_tsv = 'pseudochromosome_metadata.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df, p_df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       p_meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e4d56718-5a4d-48cf-b356-b01215304dd1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['5xCLU-h2kbKI_HO', 'CLU-h2kbKI_HO', '5xFADHEMI', '5xFADWT',\n",
       "       'hABKI_Swe_IB_HET', 'hABKI_Swe_IB_WT'], dtype=object)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df.genotype.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "398e60d7-cbe0-44ca-bd12-07d616617101",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['human_ref/pseudochrom/hCLU/CLU/CLU_fmt.gtf']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wc = {'genotype': '5xCLU-h2kbKI_HO'}\n",
    "get_df_col(wc,\n",
    "                    p_df,\n",
    "                    config['ref']['pseudochrom']['human_gene']['fmt_gtf'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a4d37d88-45d6-42c3-a5b1-43eefbef59d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_gtf = 'ref/pseudochrom/hCLU/Clu/Clu_fmt.gtf'\n",
    "human_gtf = 'human_ref/pseudochrom/hCLU/CLU/CLU_fmt.gtf'\n",
    "pseudochrom = 'hCLU'\n",
    "wc = {'pseudochrom': 'hCLU'}\n",
    "ofile = 'test_sorted_gtf.gtf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "215cb09f-97a7-457c-8ee5-fcedf56667a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge_sort_human_mouse_pseudochrom_gtfs(wc,\n",
    "                                            mouse_gtf,\n",
    "                                            human_gtf,\n",
    "                                            ofile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "28efced3-a5d2-429d-a759-23ab6706341c",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_df = pr.read_gtf(mouse_gtf).df\n",
    "h_df = pr.read_gtf(human_gtf).df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "144281f3-cb46-4fc8-9d39-19f6c772f03a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>source_attr</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>gene</td>\n",
       "      <td>3000</td>\n",
       "      <td>10510</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>TALON</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3000</td>\n",
       "      <td>15595</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>3000</td>\n",
       "      <td>3206</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>1</td>\n",
       "      <td>730959</td>\n",
       "      <td>KNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>6249</td>\n",
       "      <td>6370</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>2</td>\n",
       "      <td>730947</td>\n",
       "      <td>KNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>7842</td>\n",
       "      <td>7991</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>3</td>\n",
       "      <td>730949</td>\n",
       "      <td>KNOWN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Source     Feature  Start    End Score Strand Frame gene_id  \\\n",
       "0       hCLU  TALON        gene   3000  10510     .      +     .    hCLU   \n",
       "1       hCLU  TALON  transcript   3000  15595     .      +     .    hCLU   \n",
       "2       hCLU  TALON        exon   3000   3206     .      +     .    hCLU   \n",
       "3       hCLU  TALON        exon   6249   6370     .      +     .    hCLU   \n",
       "4       hCLU  TALON        exon   7842   7991     .      +     .    hCLU   \n",
       "\n",
       "  gene_name gene_status source_attr              transcript_id  \\\n",
       "0      hCLU       KNOWN       TALON                        NaN   \n",
       "1      hCLU       KNOWN         NaN  mouse_Clu_TALONT000141863   \n",
       "2      hCLU       KNOWN         NaN  mouse_Clu_TALONT000141863   \n",
       "3      hCLU       KNOWN         NaN  mouse_Clu_TALONT000141863   \n",
       "4      hCLU       KNOWN         NaN  mouse_Clu_TALONT000141863   \n",
       "\n",
       "  transcript_status            transcript_name exon_number exon_id exon_status  \n",
       "0               NaN                        NaN         NaN     NaN         NaN  \n",
       "1             KNOWN  mouse_Clu_TALONT000141863         NaN     NaN         NaN  \n",
       "2             KNOWN  mouse_Clu_TALONT000141863           1  730959       KNOWN  \n",
       "3             KNOWN  mouse_Clu_TALONT000141863           2  730947       KNOWN  \n",
       "4             KNOWN  mouse_Clu_TALONT000141863           3  730949       KNOWN  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "2306b736-3457-492f-b73c-6fcf98e3f088",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>source_attr</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>gene</td>\n",
       "      <td>2970</td>\n",
       "      <td>16063</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>TALON</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2970</td>\n",
       "      <td>17142</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>2970</td>\n",
       "      <td>3206</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>1</td>\n",
       "      <td>730961</td>\n",
       "      <td>KNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>6249</td>\n",
       "      <td>6370</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>2</td>\n",
       "      <td>730947</td>\n",
       "      <td>KNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>7842</td>\n",
       "      <td>7991</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>3</td>\n",
       "      <td>730949</td>\n",
       "      <td>KNOWN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Source     Feature  Start    End Score Strand Frame gene_id  \\\n",
       "0       hCLU  TALON        gene   2970  16063     .      +     .    hCLU   \n",
       "1       hCLU  TALON  transcript   2970  17142     .      +     .    hCLU   \n",
       "2       hCLU  TALON        exon   2970   3206     .      +     .    hCLU   \n",
       "3       hCLU  TALON        exon   6249   6370     .      +     .    hCLU   \n",
       "4       hCLU  TALON        exon   7842   7991     .      +     .    hCLU   \n",
       "\n",
       "  gene_name gene_status source_attr              transcript_id  \\\n",
       "0      hCLU       KNOWN       TALON                        NaN   \n",
       "1      hCLU       KNOWN         NaN  human_CLU_TALONT000141863   \n",
       "2      hCLU       KNOWN         NaN  human_CLU_TALONT000141863   \n",
       "3      hCLU       KNOWN         NaN  human_CLU_TALONT000141863   \n",
       "4      hCLU       KNOWN         NaN  human_CLU_TALONT000141863   \n",
       "\n",
       "  transcript_status            transcript_name exon_number exon_id exon_status  \n",
       "0               NaN                        NaN         NaN     NaN         NaN  \n",
       "1             KNOWN  human_CLU_TALONT000141863         NaN     NaN         NaN  \n",
       "2             KNOWN  human_CLU_TALONT000141863           1  730961       KNOWN  \n",
       "3             KNOWN  human_CLU_TALONT000141863           2  730947       KNOWN  \n",
       "4             KNOWN  human_CLU_TALONT000141863           3  730949       KNOWN  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "8dd3dbe3-7054-42a8-b2ca-f4ae82f62a27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# concat and update the \n",
    "df = pd.concat([m_df, h_df], ignore_index=True, axis=0)\n",
    "gene_entries = df.loc[df.Feature=='gene'].copy(deep=True)\n",
    "if len(gene_entries.index) == 2:\n",
    "    gene_end = gene_entries.End.max(0)\n",
    "    gene_start = gene_entries.Start.min(0)\n",
    "    df.drop(gene_entries.index.tolist()[0], axis=0, inplace=True)\n",
    "    df.loc[df.Feature=='gene', 'Start'] = gene_start\n",
    "    df.loc[df.Feature=='gene', 'End'] = gene_end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "38df0e7f-2c6e-45bf-aab2-fc6ef75737ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>source_attr</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>gene</td>\n",
       "      <td>2970</td>\n",
       "      <td>16063</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>TALON</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Source Feature  Start    End Score Strand Frame gene_id  \\\n",
       "51       hCLU  TALON    gene   2970  16063     .      +     .    hCLU   \n",
       "\n",
       "   gene_name gene_status source_attr transcript_id transcript_status  \\\n",
       "51      hCLU       KNOWN       TALON           NaN               NaN   \n",
       "\n",
       "   transcript_name exon_number exon_id exon_status  \n",
       "51             NaN         NaN     NaN         NaN  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.Feature=='gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "520df9ee-b0de-43c2-81c1-55d626c4a551",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['gene_id'] = pseudochrom\n",
    "df['gene_name'] = pseudochrom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "e80c5967-a422-4927-92c9-3c7df2eec062",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['hCLU'], dtype=object)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.gene_name.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "69b4aae6-81d5-45f3-83da-c1a6275365c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>source_attr</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3000</td>\n",
       "      <td>15595</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>3000</td>\n",
       "      <td>3206</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>1</td>\n",
       "      <td>730959</td>\n",
       "      <td>KNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>6249</td>\n",
       "      <td>6370</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>2</td>\n",
       "      <td>730947</td>\n",
       "      <td>KNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>7842</td>\n",
       "      <td>7991</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>3</td>\n",
       "      <td>730949</td>\n",
       "      <td>KNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>9360</td>\n",
       "      <td>9531</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>mouse_Clu_TALONT000141863</td>\n",
       "      <td>4</td>\n",
       "      <td>730951</td>\n",
       "      <td>KNOWN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Source     Feature  Start    End Score Strand Frame gene_id  \\\n",
       "1       hCLU  TALON  transcript   3000  15595     .      +     .    hCLU   \n",
       "2       hCLU  TALON        exon   3000   3206     .      +     .    hCLU   \n",
       "3       hCLU  TALON        exon   6249   6370     .      +     .    hCLU   \n",
       "4       hCLU  TALON        exon   7842   7991     .      +     .    hCLU   \n",
       "5       hCLU  TALON        exon   9360   9531     .      +     .    hCLU   \n",
       "\n",
       "  gene_name gene_status source_attr              transcript_id  \\\n",
       "1      hCLU       KNOWN         NaN  mouse_Clu_TALONT000141863   \n",
       "2      hCLU       KNOWN         NaN  mouse_Clu_TALONT000141863   \n",
       "3      hCLU       KNOWN         NaN  mouse_Clu_TALONT000141863   \n",
       "4      hCLU       KNOWN         NaN  mouse_Clu_TALONT000141863   \n",
       "5      hCLU       KNOWN         NaN  mouse_Clu_TALONT000141863   \n",
       "\n",
       "  transcript_status            transcript_name exon_number exon_id exon_status  \n",
       "1             KNOWN  mouse_Clu_TALONT000141863         NaN     NaN         NaN  \n",
       "2             KNOWN  mouse_Clu_TALONT000141863           1  730959       KNOWN  \n",
       "3             KNOWN  mouse_Clu_TALONT000141863           2  730947       KNOWN  \n",
       "4             KNOWN  mouse_Clu_TALONT000141863           3  730949       KNOWN  \n",
       "5             KNOWN  mouse_Clu_TALONT000141863           4  730951       KNOWN  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "27f37fdb-78bc-4709-bf13-c0ee320de78d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# try again w/ correct gtfs\n",
    "df_sorted = cerberus.sort_gtf(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "0d1ad356-2828-4d4f-8f6a-6f42321184b0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>source_attr</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>gene</td>\n",
       "      <td>2970</td>\n",
       "      <td>16063</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>TALON</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>2970</td>\n",
       "      <td>17142</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>2970</td>\n",
       "      <td>3206</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>1</td>\n",
       "      <td>730961</td>\n",
       "      <td>KNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>6249</td>\n",
       "      <td>6370</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>2</td>\n",
       "      <td>730947</td>\n",
       "      <td>KNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>7842</td>\n",
       "      <td>7991</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>human_CLU_TALONT000141863</td>\n",
       "      <td>3</td>\n",
       "      <td>730949</td>\n",
       "      <td>KNOWN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Source     Feature  Start    End Score Strand Frame gene_id  \\\n",
       "0       hCLU  TALON        gene   2970  16063     .      +     .    hCLU   \n",
       "1       hCLU  TALON  transcript   2970  17142     .      +     .    hCLU   \n",
       "2       hCLU  TALON        exon   2970   3206     .      +     .    hCLU   \n",
       "3       hCLU  TALON        exon   6249   6370     .      +     .    hCLU   \n",
       "4       hCLU  TALON        exon   7842   7991     .      +     .    hCLU   \n",
       "\n",
       "  gene_name gene_status source_attr              transcript_id  \\\n",
       "0      hCLU       KNOWN       TALON                        NaN   \n",
       "1      hCLU       KNOWN         NaN  human_CLU_TALONT000141863   \n",
       "2      hCLU       KNOWN         NaN  human_CLU_TALONT000141863   \n",
       "3      hCLU       KNOWN         NaN  human_CLU_TALONT000141863   \n",
       "4      hCLU       KNOWN         NaN  human_CLU_TALONT000141863   \n",
       "\n",
       "  transcript_status            transcript_name exon_number exon_id exon_status  \n",
       "0               NaN                        NaN         NaN     NaN         NaN  \n",
       "1             KNOWN  human_CLU_TALONT000141863         NaN     NaN         NaN  \n",
       "2             KNOWN  human_CLU_TALONT000141863           1  730961       KNOWN  \n",
       "3             KNOWN  human_CLU_TALONT000141863           2  730947       KNOWN  \n",
       "4             KNOWN  human_CLU_TALONT000141863           3  730949       KNOWN  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sorted.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "ad861073-a821-4267-b886-88933f0d861d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mouse_Clu_TALONT000141863\n",
      "mouse_Clu_TALONT000141864\n",
      "mouse_Clu_TALONT000141865\n",
      "mouse_Clu_TALONT000141866\n",
      "mouse_Clu_TALONT000141867\n",
      "mouse_Clu_TALONT000141868\n",
      "mouse_Clu_TALONT000141869\n",
      "mouse_Clu_TALONT000141870\n",
      "mouse_Clu_TALONT000141871\n",
      "nan\n",
      "human_CLU_TALONT000141863\n",
      "human_CLU_TALONT000141864\n",
      "human_CLU_TALONT000141865\n",
      "human_CLU_TALONT000141866\n",
      "human_CLU_TALONT000141867\n",
      "human_CLU_TALONT000141868\n",
      "human_CLU_TALONT000141869\n",
      "human_CLU_TALONT000141870\n",
      "human_CLU_TALONT000141871\n",
      "human_CLU_TALONT000141872\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_status</th>\n",
       "      <th>source_attr</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>exon_status</th>\n",
       "      <th>feature_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>TALON</td>\n",
       "      <td>gene</td>\n",
       "      <td>2970</td>\n",
       "      <td>16063</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>hCLU</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>TALON</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Source Feature  Start    End Score Strand Frame gene_id  \\\n",
       "51       hCLU  TALON    gene   2970  16063     .      +     .    hCLU   \n",
       "\n",
       "   gene_name gene_status source_attr transcript_id transcript_status  \\\n",
       "51      hCLU       KNOWN       TALON           NaN               NaN   \n",
       "\n",
       "   transcript_name exon_number exon_id exon_status  feature_rank  \n",
       "51             NaN         NaN     NaN         NaN             0  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for t in df.transcript_id.unique():\n",
    "    print(t)\n",
    "df.loc[df.transcript_id.isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2af82692-312e-4aaa-b155-cd2dc381494a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "98e7093e-b173-40d3-b38c-42f2f639d9e3",
   "metadata": {},
   "source": [
    "## 231107 try to make cerberus work again..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ff2069a1-11ee-4e4c-9daf-42407d94c9eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/Documents/programming/mortazavi_lab/bin/modelad_pipeline/proc_update/sm_utils.py:128: UserWarning: Config found non-pseudochrom mouse w/ genotypes ['hABKI_SWE'], is this expected?\n",
      "  warnings.warn(f'Config found non-pseudochrom mouse w/ genotypes {genotypes}, is this expected?')\n"
     ]
    }
   ],
   "source": [
    "config_tsv = '231027_config.tsv'\n",
    "# config_tsv = 'test_pseudochrom_parsing_config.tsv'\n",
    "# config_tsv = '231102_config_ad003.tsv'\n",
    "p_meta_tsv = 'pseudochromosome_metadata.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df, p_df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       p_meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "64f6b589-1d83-4feb-98d5-e55b68bf0bcc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1', '2', '4', '5', '15', '16', '17', '18', '19', '20', '21', '3',\n",
       "       '6', '7', '8', '9', '10', '11', '12', '13', '14', '22', '23', '24',\n",
       "       '25'], dtype=object)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df.cerberus_run.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ed4d1081-0430-4441-b22e-5021f089deb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "max_cerb = df.loc[df.cerberus_run.astype(int)==df.cerberus_run.astype(int).max(axis=0)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3fd07edc-f290-4884-9776-d0ae164869e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/cerberus/agg/ad008_hABKI_Swe_IB_WT_M_8_weeks_HC_25_tss.bed',\n",
       " 'data/cerberus/agg/ad008_hABKI_Swe_IB_WT_M_8_weeks_HC_25_tss.bed',\n",
       " 'data/cerberus/agg/ad008_hABKI_Swe_IB_WT_M_8_weeks_HC_25_tss.bed']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expand(expand(config['cerberus']['agg']['ends'],\n",
    "               zip,\n",
    "               study=max_cerb.study.tolist(),\n",
    "               genotype=max_cerb.genotype.tolist(),\n",
    "               sex=max_cerb.sex.tolist(),\n",
    "               age=max_cerb.age.tolist(),\n",
    "               tissue=max_cerb.tissue.tolist(),\n",
    "               cerberus_run=max_cerb.cerberus_run.tolist(),\n",
    "               allow_missing=True),\n",
    "               end_mode='tss')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "96c2263a-8bbe-4e09-8651-d5b363d5ff4a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/cerberus/agg/ad002_5xFADHEMI_M_18_months_CTX_1_tss.bed'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wc = {'cerberus_run': 2, 'end_mode': 'tss'}\n",
    "get_prev_cerb_entry(wc, p_df, config['cerberus']['agg']['ends'], config)\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ea846e63-a5ee-474c-9c2e-05866cc6f4ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/cerberus/agg/ad002_5xFADHEMI_M_18_months_CTX_1_ic.tsv'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wc = {'cerberus_run': 2}\n",
    "get_prev_cerb_entry(wc, p_df, config['cerberus']['agg']['ics'], config)\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "16ebc40d-c61e-4c19-b47e-6a1aa7c20a83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ref/ca_tss.bed'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wc = {'cerberus_run': 1, 'end_mode': 'tss'}\n",
    "get_prev_cerb_entry(wc, p_df, config['cerberus']['agg']['ends'], config)\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "36ab3f98-3260-4a9b-bd51-f3b2bdb6d391",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ref/ca_ic.tsv'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wc = {'cerberus_run': 1}\n",
    "get_prev_cerb_entry(wc, p_df, config['cerberus']['agg']['ics'], config)\n",
    "                    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fcd1b6bc-9cc7-4a33-a056-4abaf9363002",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>...</th>\n",
       "      <th>tissue</th>\n",
       "      <th>pseudochrom_needed</th>\n",
       "      <th>pseudochrom</th>\n",
       "      <th>Unnamed: 7</th>\n",
       "      <th>Unnamed: 8</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "      <th>source</th>\n",
       "      <th>cerberus_run</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad004_18876_lig-blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad004_18876_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad004_18876</td>\n",
       "      <td>18876</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>True</td>\n",
       "      <td>(hTREM2,)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hTREM2-R47HKI_HET_M_10_weeks_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>hTREM2-R47HKI_HET_M_10_weeks_HC_1</td>\n",
       "      <td>ad004_hTREM2-R47HKI_HET_M_10_weeks_HC</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad004_18876_lig-blk_2_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad004_18876_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad004_18876</td>\n",
       "      <td>18876</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>True</td>\n",
       "      <td>(hTREM2,)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hTREM2-R47HKI_HET_M_10_weeks_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>hTREM2-R47HKI_HET_M_10_weeks_HC_1</td>\n",
       "      <td>ad004_hTREM2-R47HKI_HET_M_10_weeks_HC</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     batch platform                                              fname  \\\n",
       "34  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "35  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                             basename  \\\n",
       "34  ad004_18876_lig-blk_1_t1.fastq.gz   \n",
       "35  ad004_18876_lig-blk_2_t1.fastq.gz   \n",
       "\n",
       "                                                 path flowcell  \\\n",
       "34  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "35  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "\n",
       "                file_stem  chop_num  sample_temp  mouse_id  ... tissue  \\\n",
       "34  ad004_18876_lig-blk_1       1.0  ad004_18876     18876  ...     HC   \n",
       "35  ad004_18876_lig-blk_2       1.0  ad004_18876     18876  ...     HC   \n",
       "\n",
       "   pseudochrom_needed pseudochrom Unnamed: 7 Unnamed: 8  \\\n",
       "34               True   (hTREM2,)        NaN        NaN   \n",
       "35               True   (hTREM2,)        NaN        NaN   \n",
       "\n",
       "                             sample biorep_num  \\\n",
       "34  hTREM2-R47HKI_HET_M_10_weeks_HC          1   \n",
       "35  hTREM2-R47HKI_HET_M_10_weeks_HC          1   \n",
       "\n",
       "                              dataset                                 source  \\\n",
       "34  hTREM2-R47HKI_HET_M_10_weeks_HC_1  ad004_hTREM2-R47HKI_HET_M_10_weeks_HC   \n",
       "35  hTREM2-R47HKI_HET_M_10_weeks_HC_1  ad004_hTREM2-R47HKI_HET_M_10_weeks_HC   \n",
       "\n",
       "   cerberus_run  \n",
       "34            9  \n",
       "35            9  \n",
       "\n",
       "[2 rows x 24 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max_cerb = df.loc[df.cerberus_run==df.cerberus_run.max(axis=0)]\n",
    "max_cerb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "bde7dede-b1a3-4ee7-826e-5817415f8b77",
   "metadata": {},
   "outputs": [],
   "source": [
    "ics = expand(config['cerberus']['ics'], \n",
    "    zip,\n",
    "    study=df.study.tolist(),\n",
    "    genotype=df.genotype.tolist(),\n",
    "    sex=df.sex.tolist(),\n",
    "    age=df.age.tolist(),\n",
    "    tissue=df.tissue.tolist(),\n",
    "    cerberus_run=df.cerberus_run.tolist()) \n",
    "tsss = expand(expand(config['cerberus']['ends'], \n",
    "    zip,\n",
    "    study=df.study.tolist(),\n",
    "    genotype=df.genotype.tolist(),\n",
    "    sex=df.sex.tolist(),\n",
    "    age=df.age.tolist(),\n",
    "    tissue=df.tissue.tolist(),\n",
    "    cerberus_run=df.cerberus_run.tolist(),\n",
    "    allow_missing=True),\n",
    "    end_mode='tss') \n",
    "tess = expand(expand(config['cerberus']['ends'], \n",
    "    zip,\n",
    "    study=df.study.tolist(),\n",
    "    genotype=df.genotype.tolist(),\n",
    "    sex=df.sex.tolist(),\n",
    "    age=df.age.tolist(),\n",
    "    tissue=df.tissue.tolist(),\n",
    "    cerberus_run=df.cerberus_run.tolist(),\n",
    "    allow_missing=True),\n",
    "    end_mode='tes') \n",
    "                                "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d4dedd82-06b2-474f-ae2b-0068029d9617",
   "metadata": {},
   "outputs": [],
   "source": [
    "for f in ics+tsss+tess:\n",
    "    os.makedirs(os.path.dirname(f), exist_ok=True)\n",
    "    pathlib.Path(f).touch()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "cd1dd237-bd36-4ea5-9e25-6cf81a4e6375",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['5xFADHEMI',\n",
       " '5xFADWT',\n",
       " '5xCLU-h2kbKI_HO',\n",
       " 'CLU-h2kbKI_HO',\n",
       " 'hTREM2-R47HKI_HET',\n",
       " 'hTREM2_KI_HET',\n",
       " 'hTREM2_KI_WT',\n",
       " '5xBIN1_HO',\n",
       " 'BIN1_HO',\n",
       " 'C57B6J',\n",
       " 'hABKI_SWE',\n",
       " 'ABCA7_HO',\n",
       " 'hABKI_Swe_IB_HET',\n",
       " 'hABKI_Swe_IB_WT']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.genotype.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f537032-c996-4fee-aaa3-9573e9cc02a7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afb74b03-4357-4afb-9006-0936d71f8f7e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b58ba286-cab8-43d7-963a-c5c78b3be359",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bcdcd5fe-576d-4ebc-8b67-5ad622b6dd8d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24b4e2c0-d246-4e36-8d7d-90bd77d33335",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "73b70cfe-c702-48ec-987f-42023f4874b5",
   "metadata": {},
   "source": [
    "## 231103 request all output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "2d95ab86-9164-4f4c-b549-033b37b466eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ref/pseudochrom/{pseudochrom}/ref.fa'"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config['ref']['pseudochrom']['fa']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "2e8f78f6-a48b-43cb-91df-e80115776a40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# config_tsv = '231020_config.tsv'\n",
    "# config_tsv = 'test_pseudochrom_parsing_config.tsv'\n",
    "config_tsv = '231102_config_ad003.tsv'\n",
    "p_meta_tsv = 'pseudochromosome_metadata.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df, p_df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       p_meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "17ac66a9-f6b2-4124-9911-4044591028d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pseudochrom</th>\n",
       "      <th>mouse_gene</th>\n",
       "      <th>human_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>Clu</td>\n",
       "      <td>CLU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>dummy</td>\n",
       "      <td>dummy</td>\n",
       "      <td>dummy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pseudochrom mouse_gene human_gene\n",
       "0         hCLU        Clu        CLU\n",
       "12       dummy      dummy      dummy"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df[['pseudochrom', 'mouse_gene', 'human_gene']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "27effa34-609e-44a2-9841-434233fd53d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ref/pseudochrom/dummy/dummy/config.csv',\n",
       " 'ref/pseudochrom/hCLU/Clu/config.csv']"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(set(expand(config['ref']['pseudochrom']['gene']['config'],\n",
    "       zip,\n",
    "       pseudochrom=p_df.pseudochrom.tolist(),\n",
    "       mouse_gene=p_df.mouse_gene.tolist())))\n",
    "list(set(expand(config['ref']['pseudochrom']['human_gene']['config'],\n",
    "       zip,\n",
    "       pseudochrom=p_df.pseudochrom.tolist(),\n",
    "       human_gene=p_df.human_gene.tolist())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "860d9f49-09a0-4af5-886a-09e1c1f1e5a6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['human_ref/pseudochrom/hCLU/CLU/config.csv',\n",
       " 'human_ref/pseudochrom/dummy/dummy/config.csv']"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(set(expand(config['ref']['pseudochrom']['human_gene']['config'],\n",
    "       zip,\n",
    "       pseudochrom=p_df.pseudochrom.tolist(),\n",
    "       human_gene=p_df.human_gene.tolist())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5bd7770-e6f6-49a1-b305-bc00fc74f407",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref/pseudochrom/{pseudochrom}/{mouse_gene}/config.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e42775bc-7429-432d-a8dd-2899323ce44a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>study</th>\n",
       "      <th>...</th>\n",
       "      <th>Unnamed: 8</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "      <th>source</th>\n",
       "      <th>cerberus_run</th>\n",
       "      <th>human_gene</th>\n",
       "      <th>mouse_gene</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12517_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_12517_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12517</td>\n",
       "      <td>12517</td>\n",
       "      <td>ad003</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC_1</td>\n",
       "      <td>ad003_5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>CLU</td>\n",
       "      <td>Clu</td>\n",
       "      <td>chimeric</td>\n",
       "      <td>Some full exons are human; some are mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12517_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_12517_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12517</td>\n",
       "      <td>12517</td>\n",
       "      <td>ad003</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC_1</td>\n",
       "      <td>ad003_5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>CLU</td>\n",
       "      <td>Clu</td>\n",
       "      <td>chimeric</td>\n",
       "      <td>Some full exons are human; some are mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12649_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_12649_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12649</td>\n",
       "      <td>12649</td>\n",
       "      <td>ad003</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC_2</td>\n",
       "      <td>ad003_5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>CLU</td>\n",
       "      <td>Clu</td>\n",
       "      <td>chimeric</td>\n",
       "      <td>Some full exons are human; some are mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12649_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_12649_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12649</td>\n",
       "      <td>12649</td>\n",
       "      <td>ad003</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC_2</td>\n",
       "      <td>ad003_5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>CLU</td>\n",
       "      <td>Clu</td>\n",
       "      <td>chimeric</td>\n",
       "      <td>Some full exons are human; some are mouse</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12659_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_12659_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12659</td>\n",
       "      <td>12659</td>\n",
       "      <td>ad003</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC_3</td>\n",
       "      <td>ad003_5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>CLU</td>\n",
       "      <td>Clu</td>\n",
       "      <td>chimeric</td>\n",
       "      <td>Some full exons are human; some are mouse</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  platform                                              fname  \\\n",
       "0      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "1      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "2      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "3      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "4      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                         basename  \\\n",
       "0  ad003_12517_lig-blk_1_t1.fastq   \n",
       "1  ad003_12517_lig-blk_2_t1.fastq   \n",
       "2  ad003_12649_lig-blk_1_t1.fastq   \n",
       "3  ad003_12649_lig-blk_2_t1.fastq   \n",
       "4  ad003_12659_lig-blk_1_t1.fastq   \n",
       "\n",
       "                                                path flowcell  \\\n",
       "0  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "1  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "2  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "3  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "4  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "\n",
       "               file_stem  chop_num  sample_temp  mouse_id  study  ...  \\\n",
       "0  ad003_12517_lig-blk_1       1.0  ad003_12517     12517  ad003  ...   \n",
       "1  ad003_12517_lig-blk_2       1.0  ad003_12517     12517  ad003  ...   \n",
       "2  ad003_12649_lig-blk_1       1.0  ad003_12649     12649  ad003  ...   \n",
       "3  ad003_12649_lig-blk_2       1.0  ad003_12649     12649  ad003  ...   \n",
       "4  ad003_12659_lig-blk_1       1.0  ad003_12659     12659  ad003  ...   \n",
       "\n",
       "  Unnamed: 8                         sample biorep_num  \\\n",
       "0        NaN  5xCLU-h2kbKI_HO_F_4_months_HC          1   \n",
       "1        NaN  5xCLU-h2kbKI_HO_F_4_months_HC          1   \n",
       "2        NaN  5xCLU-h2kbKI_HO_F_4_months_HC          2   \n",
       "3        NaN  5xCLU-h2kbKI_HO_F_4_months_HC          2   \n",
       "4        NaN  5xCLU-h2kbKI_HO_F_4_months_HC          3   \n",
       "\n",
       "                           dataset                               source  \\\n",
       "0  5xCLU-h2kbKI_HO_F_4_months_HC_1  ad003_5xCLU-h2kbKI_HO_F_4_months_HC   \n",
       "1  5xCLU-h2kbKI_HO_F_4_months_HC_1  ad003_5xCLU-h2kbKI_HO_F_4_months_HC   \n",
       "2  5xCLU-h2kbKI_HO_F_4_months_HC_2  ad003_5xCLU-h2kbKI_HO_F_4_months_HC   \n",
       "3  5xCLU-h2kbKI_HO_F_4_months_HC_2  ad003_5xCLU-h2kbKI_HO_F_4_months_HC   \n",
       "4  5xCLU-h2kbKI_HO_F_4_months_HC_3  ad003_5xCLU-h2kbKI_HO_F_4_months_HC   \n",
       "\n",
       "  cerberus_run  human_gene  mouse_gene locus_type  \\\n",
       "0            1         CLU         Clu   chimeric   \n",
       "1            1         CLU         Clu   chimeric   \n",
       "2            1         CLU         Clu   chimeric   \n",
       "3            1         CLU         Clu   chimeric   \n",
       "4            1         CLU         Clu   chimeric   \n",
       "\n",
       "                                       notes  \n",
       "0  Some full exons are human; some are mouse  \n",
       "1  Some full exons are human; some are mouse  \n",
       "2  Some full exons are human; some are mouse  \n",
       "3  Some full exons are human; some are mouse  \n",
       "4  Some full exons are human; some are mouse  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "230bb54f-d2a6-40a9-93ad-31a8e9dfb039",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45352cad-be2d-4934-9d3f-107b41b19889",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b5b3dc1-181e-4ee1-9f32-be89edd3c200",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7d7b102a-d6ab-4126-a407-dd74ba0753cf",
   "metadata": {},
   "source": [
    "## 231103 get output files which are the mapped human / mouse genes to the pseudochromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ea884dd9-7d02-4d5c-a8f8-2640c9b946bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# config_tsv = '231020_config.tsv'\n",
    "# config_tsv = 'test_pseudochrom_parsing_config.tsv'\n",
    "config_tsv = '231102_config_ad003.tsv'\n",
    "p_meta_tsv = 'pseudochromosome_metadata.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df, p_df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       p_meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "67305052-dd41-4ab1-85c6-8d9246691d88",
   "metadata": {},
   "outputs": [],
   "source": [
    "db = config['ref']['talon']['db']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f3a326b3-c4cd-4a54-b447-8b8867c6aac6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mk_pseudochrom_mapped_gene_talon_config(wc,\n",
    "                                            species,\n",
    "                                            p_df,\n",
    "                                            ofile):    \n",
    "    if species == 'mouse':\n",
    "        cfg_entry = config['ref']['pseudochrom']['gene']['sort_bam']\n",
    "        col = 'mouse_gene'\n",
    "    elif species == 'human':\n",
    "        cfg_entry = config['ref']['pseudochrom']['human_gene']['sort_bam']\n",
    "        col = 'human_gene'\n",
    "    gene = get_df_col(wc, p_df, col)\n",
    "    \n",
    "    if gene == 'dummy':\n",
    "        pathlib.Path(ofile).touch()\n",
    "\n",
    "    else:\n",
    "        # input files\n",
    "        bam = get_cfg_entries(wc, \n",
    "                              p_df,\n",
    "                              cfg_entry)\n",
    "        print(bam)\n",
    "\n",
    "        # actual config file makery\n",
    "        temp = get_cfg_entries(wildcards,\n",
    "                                   p_df,\n",
    "                                   cfg_entry,\n",
    "                                   return_df=True)\n",
    "        print(gene)\n",
    "        print(temp)\n",
    "        temp = temp[['sample', 'dataset', 'platform', 'file']]\n",
    "        # temp_str = f'{species}_{gene}'\n",
    "        # temp['sample'] = temp_str\n",
    "        # temp['dataset'] = temp_str\n",
    "        temp['sample'] = gene\n",
    "        temp['dataset'] = gene\n",
    "        print(temp)\n",
    "        temp.to_csv(ofile, index=False, header=None, sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fb39921c-45a4-4dda-9302-e8dac3ecd959",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ref/pseudochrom/hCLU/Clu/mapped_sorted.bam']\n",
      "Clu\n",
      "  platform                                              fname  \\\n",
      "0      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
      "\n",
      "                         basename  \\\n",
      "0  ad003_12517_lig-blk_1_t1.fastq   \n",
      "\n",
      "                                                path flowcell  \\\n",
      "0  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
      "\n",
      "               file_stem  chop_num  sample_temp  mouse_id  study  ...  \\\n",
      "0  ad003_12517_lig-blk_1       1.0  ad003_12517     12517  ad003  ...   \n",
      "\n",
      "                          sample biorep_num                          dataset  \\\n",
      "0  5xCLU-h2kbKI_HO_F_4_months_HC          1  5xCLU-h2kbKI_HO_F_4_months_HC_1   \n",
      "\n",
      "                                source  cerberus_run human_gene  mouse_gene  \\\n",
      "0  ad003_5xCLU-h2kbKI_HO_F_4_months_HC             1        CLU         Clu   \n",
      "\n",
      "   locus_type                                      notes  \\\n",
      "0    chimeric  Some full exons are human; some are mouse   \n",
      "\n",
      "                                         file  \n",
      "0  ref/pseudochrom/hCLU/Clu/mapped_sorted.bam  \n",
      "\n",
      "[1 rows x 28 columns]\n",
      "  sample dataset platform                                        file\n",
      "0    Clu     Clu      ONT  ref/pseudochrom/hCLU/Clu/mapped_sorted.bam\n"
     ]
    }
   ],
   "source": [
    "wildcards = {'genotype': '5xCLU-h2kbKI_HO', 'mouse_gene': 'Clu'}\n",
    "species = 'mouse'\n",
    "\n",
    "mk_pseudochrom_mapped_gene_talon_config(wildcards,\n",
    "                                        species,\n",
    "                                        p_df,\n",
    "                                        'test_talon_cfg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4e662cdf-7b38-4a5e-a1b8-27066d0fb378",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ref/pseudochrom/dummy/dummy/mapped_sorted.bam']\n",
      "dummy\n",
      "   platform                                              fname  \\\n",
      "18      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
      "\n",
      "                          basename  \\\n",
      "18  ad003_11627_lig-blk_1_t1.fastq   \n",
      "\n",
      "                                                 path flowcell  \\\n",
      "18  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
      "\n",
      "                file_stem  chop_num  sample_temp  mouse_id  study  ...  \\\n",
      "18  ad003_11627_lig-blk_1       1.0  ad003_11627     11627  ad003  ...   \n",
      "\n",
      "                   sample biorep_num                  dataset  \\\n",
      "18  5xFADWT_F_4_months_HC          1  5xFADWT_F_4_months_HC_1   \n",
      "\n",
      "                         source  cerberus_run human_gene  mouse_gene  \\\n",
      "18  ad003_5xFADWT_F_4_months_HC             3      dummy       dummy   \n",
      "\n",
      "    locus_type  notes                                           file  \n",
      "18       dummy  dummy  ref/pseudochrom/dummy/dummy/mapped_sorted.bam  \n",
      "\n",
      "[1 rows x 28 columns]\n",
      "   sample dataset platform                                           file\n",
      "18  dummy   dummy      ONT  ref/pseudochrom/dummy/dummy/mapped_sorted.bam\n"
     ]
    }
   ],
   "source": [
    "wildcards = {'genotype': '5xFADWT', 'mouse_gene': 'dummy'}\n",
    "species = 'mouse'\n",
    "\n",
    "mk_pseudochrom_mapped_gene_talon_config(wildcards,\n",
    "                                        species,\n",
    "                                        p_df,\n",
    "                                        'test_talon_cfg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "84eea908-ec76-4526-8b43-d93e5f7fc9b5",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'tolist'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-17-366d94c95400>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0;31m# actual config file makery\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 20\u001b[0;31m \u001b[0mgene\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_df_col\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcol\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     21\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m temp = get_cfg_entries(wildcards,\n",
      "\u001b[0;32m~/Documents/programming/mortazavi_lab/bin/modelad_pipeline/proc_update/sm_utils.py\u001b[0m in \u001b[0;36mget_df_col\u001b[0;34m(wc, df, col, allow_multiple)\u001b[0m\n\u001b[1;32m    222\u001b[0m                 \u001b[0mmsg\u001b[0m\u001b[0;34m+=\u001b[0m\u001b[0;34mf'\\n{key}: {item}'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    223\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 224\u001b[0;31m         \u001b[0mval\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    225\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    226\u001b[0m         \u001b[0mval\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5485\u001b[0m         ):\n\u001b[1;32m   5486\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5487\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5488\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5489\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'tolist'"
     ]
    }
   ],
   "source": [
    "# for a given genotype, get the human and mouse gene mappings to \n",
    "# be given to talon\n",
    "wildcards = {'genotype': '5xCLU-h2kbKI_HO'}\n",
    "species = 'mouse'\n",
    "\n",
    "if species == 'mouse':\n",
    "    cfg_entry = config['ref']['pseudochrom']['gene']['sort_bam']\n",
    "    col = 'mouse_gene'\n",
    "elif species == 'human':\n",
    "    cfg_entry = config['ref']['pseudochrom']['human_gene']['sort_bam']\n",
    "    col = 'human_gene'\n",
    "    \n",
    "# input files\n",
    "bam = get_cfg_entries(wc, \n",
    "                     p_df,\n",
    "                     cfg_entry)\n",
    "db = config['ref']['talon']['db']\n",
    "\n",
    "# actual config file makery\n",
    "gene = get_df_col(wc, p_df, col)\n",
    "print(gene)\n",
    "temp = get_cfg_entries(wildcards,\n",
    "                       p_df,\n",
    "                       cfg_entry,\n",
    "                       return_df=True)\n",
    "# print(bam)\n",
    "# print(gene)\n",
    "# print(temp)\n",
    "temp = temp[['sample', 'dataset', 'platform', 'file']]\n",
    "# temp_str = f'{species}_{gene}'\n",
    "# temp['sample'] = temp_str\n",
    "# temp['dataset'] = temp_str\n",
    "temp['sample'] = gene\n",
    "temp['dataset'] = gene\n",
    "temp.to_csv(output.cfg, index=False, header=None, sep=',')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "d5acfe19-f344-4647-80f9-d42ac0fbdb4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ref/pseudochrom/hCLU/Clu/mapped_sorted.bam']\n",
      "Clu\n",
      "['human_ref/pseudochrom/hCLU/CLU/mapped_sorted.bam']\n",
      "CLU\n"
     ]
    }
   ],
   "source": [
    "# for a given genotype, get the human and mouse gene mappings to \n",
    "# be given to talon\n",
    "wildcards = {'genotype': '5xCLU-h2kbKI_HO'}\n",
    "\n",
    "# input files\n",
    "mouse_bam = get_cfg_entries(wc, \n",
    "                     p_df,\n",
    "                     config['ref']['gene']['sort_bam'])\n",
    "human_bam = get_cfg_entries(wc, \n",
    "                     p_df,\n",
    "                     config['human_ref']['gene']['sort_bam'])\n",
    "db = config['ref']['talon']['db']\n",
    "\n",
    "# actual config file makery\n",
    "mouse_gene = get_df_col(wc, \n",
    "                 p_df, \n",
    "                 'mouse_gene')\n",
    "human_gene = get_df_col(wc, p_df, 'human_gene')\n",
    "\n",
    "\n",
    "print(mouse_bam)\n",
    "print(mouse_gene)\n",
    "\n",
    "print(human_bam)\n",
    "print(human_gene)\n",
    "\n",
    "# if both entries are dummy, just \n",
    "\n",
    "# if human_gene != 'dummy':\n",
    "#     temp = get_cfg_entries(wildcards,\n",
    "#                            p_df,\n",
    "#                            config['human_ref']['gene']['sort_bam'],\n",
    "#                            return_df=True)\n",
    "# else: \n",
    "#     temp = pd.DataFrame()\n",
    "\n",
    "# if mouse_gene != 'dummy':\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "86e05b0e-983b-4521-bf38-2e4677d4f337",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ref/talon/ref.db'"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "5f3778fa-bc90-46a1-a7ff-3e929a8c6d7c",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'tolist'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-13-b45347c429ca>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      7\u001b[0m eep = get_df_col(wc, \n\u001b[1;32m      8\u001b[0m                  \u001b[0mp_df\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m                  'mouse_gene')\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0meep\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/programming/mortazavi_lab/bin/modelad_pipeline/proc_update/sm_utils.py\u001b[0m in \u001b[0;36mget_df_col\u001b[0;34m(wc, df, col, allow_multiple)\u001b[0m\n\u001b[1;32m    222\u001b[0m                 \u001b[0mmsg\u001b[0m\u001b[0;34m+=\u001b[0m\u001b[0;34mf'\\n{key}: {item}'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    223\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmsg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 224\u001b[0;31m         \u001b[0mval\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    225\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    226\u001b[0m         \u001b[0mval\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcol\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5485\u001b[0m         ):\n\u001b[1;32m   5486\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5487\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5488\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5489\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'tolist'"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "# wc = {'pseudochrom': 'hCLU'}\n",
    "\n",
    "# input files\n",
    "\n",
    "\n",
    "wc = {'genotype': '5xFADWT'}\n",
    "eep = get_df_col(wc, \n",
    "                 p_df, \n",
    "                 'mouse_gene')\n",
    "print(eep)\n",
    "\n",
    "eep = get_cfg_entries(wc, \n",
    "                     p_df,\n",
    "                     config['ref']['pseudochrom']['gene']['sort_bam'])\n",
    "print(eep)\n",
    "eep = get_cfg_entries(wc, \n",
    "                     p_df,\n",
    "                     config['ref']['pseudochrom']['human_gene']['sort_bam'])\n",
    "print(eep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f8c02eb-3c03-4a91-a7ec-de260002e549",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "711cfb34-2b48-4a50-b3f0-ed0c6c9de200",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "097a8602-70c0-4735-aa4b-ce327a1bb36c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6917a7b6-5315-4c33-a778-23f607e90006",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8a6340a-7bd0-4a0f-83d3-ff9face2ad2b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "427230b2-2b4a-4cf4-ba1f-78a43e36b3c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['5xCLU-h2kbKI_HO',\n",
       " '5xCLU-h2kbKI_HO',\n",
       " '5xCLU-h2kbKI_HO',\n",
       " '5xCLU-h2kbKI_HO',\n",
       " '5xCLU-h2kbKI_HO',\n",
       " '5xCLU-h2kbKI_HO',\n",
       " 'CLU-h2kbKI_HO',\n",
       " 'CLU-h2kbKI_HO',\n",
       " 'CLU-h2kbKI_HO',\n",
       " 'CLU-h2kbKI_HO',\n",
       " 'CLU-h2kbKI_HO',\n",
       " 'CLU-h2kbKI_HO',\n",
       " '5xFADHEMI',\n",
       " '5xFADHEMI',\n",
       " '5xFADHEMI',\n",
       " '5xFADHEMI',\n",
       " '5xFADHEMI',\n",
       " '5xFADHEMI',\n",
       " '5xFADWT',\n",
       " '5xFADWT',\n",
       " '5xFADWT',\n",
       " '5xFADWT',\n",
       " '5xFADWT',\n",
       " '5xFADWT']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df.genotype.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4f8b899-2154-41e8-9bc9-6e9984b2e639",
   "metadata": {},
   "outputs": [],
   "source": [
    "expand(config['ref']['gene']['ind_bam'],\n",
    "       pseudochrom="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a18ad540-4b75-4ba8-b46e-da711a25c1fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pseudochrom</th>\n",
       "      <th>human_gene</th>\n",
       "      <th>mouse_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hCLU</td>\n",
       "      <td>CLU</td>\n",
       "      <td>Clu</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>dummy</td>\n",
       "      <td>dummy</td>\n",
       "      <td>dummy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pseudochrom human_gene mouse_gene\n",
       "0         hCLU        CLU        Clu\n",
       "12       dummy      dummy      dummy"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_df[['pseudochrom', 'human_gene', 'mouse_gene']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9a2ea22f-e67f-4656-a47a-3b83dc8a5331",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ref/pseudochrom/hCLU/Clu/mapped_sorted.bam']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wc = {'pseudochrom': 'hCLU'}\n",
    "eep = get_cfg_entries(wc, \n",
    "                     p_df,\n",
    "                     config['ref']['gene']['sort_bam'])\n",
    "eep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "e04161f3-b609-487a-b014-40d3fbdee312",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['human_ref/pseudochrom/hCLU/CLU/mapped_sorted.bam']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wc = {'pseudochrom': 'hCLU'}\n",
    "eep = get_cfg_entries(wc, \n",
    "                     p_df,\n",
    "                     config['human_ref']['gene']['sort_bam'])\n",
    "eep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b66fbc16-e10f-4789-97a5-62fbc499f998",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'human_gene'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-15-a86e03b9f469>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mwildcards\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0;34m'genotype'\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m'5xCLU-h2kbKI_HO'\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mget_cfg_entries\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwildcards\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconfig\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'merge'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sort_bam'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreturn_df\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mtemp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sample'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'dataset'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'platform'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'file'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcfg\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m','\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/Documents/programming/mortazavi_lab/bin/modelad_pipeline/proc_update/sm_utils.py\u001b[0m in \u001b[0;36mget_cfg_entries\u001b[0;34m(wc, df, cfg_entry, return_df)\u001b[0m\n\u001b[1;32m    249\u001b[0m     \u001b[0mcerberus_run\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcerberus_run\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    250\u001b[0m     \u001b[0mpseudochrom\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpseudochrom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 251\u001b[0;31m     \u001b[0mhuman_gene\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhuman_gene\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    252\u001b[0m     \u001b[0mmouse_gene\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmouse_gene\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    253\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5485\u001b[0m         ):\n\u001b[1;32m   5486\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5487\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5488\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5489\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m->\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'human_gene'"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca12e7df-64a6-48a6-a767-fa54ea428154",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f79cfd54-7c29-4516-993a-b4e21bc46665",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edba5982-5b12-4025-935c-53feb6389f4e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "7fd422d7-71b8-4e69-bfa8-6b9cc0ec157d",
   "metadata": {},
   "source": [
    "## 231102 get input files for each genotype's reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "64e6493b-f6e9-472f-a94d-0c040967c274",
   "metadata": {},
   "outputs": [],
   "source": [
    "# config_tsv = '231020_config.tsv'\n",
    "# config_tsv = 'test_pseudochrom_parsing_config.tsv'\n",
    "config_tsv = '231102_config_ad003.tsv'\n",
    "p_meta_tsv = 'pseudochromosome_metadata.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df, p_df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       p_meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3604b6d6-7448-4676-9828-21b29910a621",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['hCLU']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wc = {'genotype': '5xCLU-h2kbKI_HO'}\n",
    "get_df_col(wc, \n",
    "           p_df, \n",
    "           'pseudochrom',\n",
    "           allow_multiple=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "707a7a56-25cb-40dd-8887-cd7d26c14810",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "['dummy'] == get_df_col(wc, \n",
    "           p_df, \n",
    "           'pseudochrom',\n",
    "            allow_multiple=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3b564885-e843-4814-8a7d-40f73aab648d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['5xCLU-h2kbKI_HO', '5xFADHEMI', '5xFADWT', 'CLU-h2kbKI_HO'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.genotype.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4fb79ace-39d4-4f0b-adf2-fe7065db0430",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['dummy']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wc = {'genotype': '5xFADHEMI'}\n",
    "get_df_col(wc, \n",
    "           p_df, \n",
    "           'pseudochrom',\n",
    "           allow_multiple=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a1dea929-dbec-4ae0-9753-1649ad2e8e5b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "['dummy'] == get_df_col(wc, \n",
    "           p_df, \n",
    "           'pseudochrom',\n",
    "            allow_multiple=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "6cc62779-1b5f-4333-a821-38d0771c9d3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# config_tsv = '231020_config.tsv'\n",
    "config_tsv = 'test_pseudochrom_parsing_config.tsv'\n",
    "# config_tsv = '231102_config_ad003.tsv'\n",
    "p_meta_tsv = 'pseudochromosome_metadata.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df, p_df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       p_meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b34fd4f7-ab84-4d19-9477-eced2396d71a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['hgenotype1', 'hgenotype2']"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wc = {'genotype': ('hgenotype1_genotype2')}\n",
    "get_df_col(wc, \n",
    "           p_df, \n",
    "           'pseudochrom',\n",
    "           allow_multiple=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4fc29845-cb1c-4a4b-bb94-c91890b5198a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "['dummy'] == get_df_col(wc, \n",
    "           p_df, \n",
    "           'pseudochrom',\n",
    "           allow_multiple=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "c4485c02-bf3b-4800-97c7-7861d7c5fcf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CLU-h2kbKI_HO', 'hTREM2_KI_HET', 'hTREM2_KI_WT',\n",
       "       'hgenotype1_genotype2'], dtype=object)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7e041e20-508b-4b22-b3dc-1295a9256628",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get all the pseudo chromosomes that are needed \n",
    "# need for the given genotype\n",
    "wc = {'genotype': ('hgenotype1_genotype2')}\n",
    "eep = get_cfg_entries(wc, \n",
    "     p_df,\n",
    "     config['ref']['pseudochrom']['fa'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "593d9f8c-9030-45e6-ba48-ecf412fab614",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ref/pseudochrom/hgenotype1/ref.fa', 'ref/pseudochrom/hgenotype2/ref.fa']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eep"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "55fcf227-6118-4aa8-b299-18628351da59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ref/genome/hgenotype1_genotype2/merged.fa']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# output\n",
    "get_cfg_entries(wc, \n",
    "                df,\n",
    "                config['ref']['pseudochrom']['fa_merge'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fb6bfd3-9cc7-496f-b677-fb5556caa251",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19485632-3e61-4106-8f12-d38e336252d2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cfb98302-0f48-45cc-86cb-6fd00279a60a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f662139a-59cb-40d6-9d0e-d97356bc7d77",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d5d630fb-099b-4e08-9a18-e924775e197b",
   "metadata": {},
   "source": [
    "## 231102 adding a pseudochrom metadata table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c8a8b2be-b4c5-419a-b070-4f8f5a04cb84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# config_tsv = '231020_config.tsv'\n",
    "config_tsv = 'test_pseudochrom_parsing_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1b9e9272-a4e4-402a-8a22-5359aaee3bdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>...</th>\n",
       "      <th>tissue</th>\n",
       "      <th>pseudochrom_needed</th>\n",
       "      <th>pseudochrom</th>\n",
       "      <th>human_gene</th>\n",
       "      <th>mouse_gene</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "      <th>source</th>\n",
       "      <th>cerberus_run</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12670_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_12670_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12670</td>\n",
       "      <td>12670</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>True</td>\n",
       "      <td>(hCLU,)</td>\n",
       "      <td>CLU</td>\n",
       "      <td>Clu</td>\n",
       "      <td>CLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>CLU-h2kbKI_HO_F_4_months_HC_1</td>\n",
       "      <td>ad003_CLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12670_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_12670_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12670</td>\n",
       "      <td>12670</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>True</td>\n",
       "      <td>(hCLU,)</td>\n",
       "      <td>CLU</td>\n",
       "      <td>Clu</td>\n",
       "      <td>CLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>CLU-h2kbKI_HO_F_4_months_HC_1</td>\n",
       "      <td>ad003_CLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad004_18634_lig-blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad004_18634_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad004_18634</td>\n",
       "      <td>18634</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>True</td>\n",
       "      <td>(hTREM2,)</td>\n",
       "      <td>TREM2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hTREM2_KI_HET_M_8_weeks_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>hTREM2_KI_HET_M_8_weeks_HC_1</td>\n",
       "      <td>ad004_hTREM2_KI_HET_M_8_weeks_HC</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad004_18503_lig-blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad004_18503_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad004_18503</td>\n",
       "      <td>18503</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>True</td>\n",
       "      <td>(hTREM2,)</td>\n",
       "      <td>TREM2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hTREM2_KI_WT_M_8_weeks_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>hTREM2_KI_WT_M_8_weeks_HC_1</td>\n",
       "      <td>ad004_hTREM2_KI_WT_M_8_weeks_HC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad004_18503_lig-blk_2_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad004_18503_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad004_18503</td>\n",
       "      <td>18503</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>True</td>\n",
       "      <td>(hTREM2,)</td>\n",
       "      <td>TREM2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>hTREM2_KI_WT_M_8_weeks_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>hTREM2_KI_WT_M_8_weeks_HC_1</td>\n",
       "      <td>ad004_hTREM2_KI_WT_M_8_weeks_HC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    batch platform                                              fname  \\\n",
       "0  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "1  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "2  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "3  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "4  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                            basename  \\\n",
       "0     ad003_12670_lig-blk_1_t1.fastq   \n",
       "1     ad003_12670_lig-blk_2_t1.fastq   \n",
       "2  ad004_18634_lig-blk_1_t1.fastq.gz   \n",
       "3  ad004_18503_lig-blk_1_t1.fastq.gz   \n",
       "4  ad004_18503_lig-blk_2_t1.fastq.gz   \n",
       "\n",
       "                                                path flowcell  \\\n",
       "0  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "1  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "2  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "3  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "4  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "\n",
       "               file_stem  chop_num  sample_temp  mouse_id  ... tissue  \\\n",
       "0  ad003_12670_lig-blk_1       1.0  ad003_12670     12670  ...     HC   \n",
       "1  ad003_12670_lig-blk_2       1.0  ad003_12670     12670  ...     HC   \n",
       "2  ad004_18634_lig-blk_1       1.0  ad004_18634     18634  ...     HC   \n",
       "3  ad004_18503_lig-blk_1       1.0  ad004_18503     18503  ...     HC   \n",
       "4  ad004_18503_lig-blk_2       1.0  ad004_18503     18503  ...     HC   \n",
       "\n",
       "  pseudochrom_needed pseudochrom human_gene mouse_gene  \\\n",
       "0               True     (hCLU,)        CLU        Clu   \n",
       "1               True     (hCLU,)        CLU        Clu   \n",
       "2               True   (hTREM2,)      TREM2        NaN   \n",
       "3               True   (hTREM2,)      TREM2        NaN   \n",
       "4               True   (hTREM2,)      TREM2        NaN   \n",
       "\n",
       "                        sample biorep_num                        dataset  \\\n",
       "0  CLU-h2kbKI_HO_F_4_months_HC          1  CLU-h2kbKI_HO_F_4_months_HC_1   \n",
       "1  CLU-h2kbKI_HO_F_4_months_HC          1  CLU-h2kbKI_HO_F_4_months_HC_1   \n",
       "2   hTREM2_KI_HET_M_8_weeks_HC          1   hTREM2_KI_HET_M_8_weeks_HC_1   \n",
       "3    hTREM2_KI_WT_M_8_weeks_HC          1    hTREM2_KI_WT_M_8_weeks_HC_1   \n",
       "4    hTREM2_KI_WT_M_8_weeks_HC          1    hTREM2_KI_WT_M_8_weeks_HC_1   \n",
       "\n",
       "                              source cerberus_run  \n",
       "0  ad003_CLU-h2kbKI_HO_F_4_months_HC            1  \n",
       "1  ad003_CLU-h2kbKI_HO_F_4_months_HC            1  \n",
       "2   ad004_hTREM2_KI_HET_M_8_weeks_HC            2  \n",
       "3    ad004_hTREM2_KI_WT_M_8_weeks_HC            3  \n",
       "4    ad004_hTREM2_KI_WT_M_8_weeks_HC            3  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "76773fcf-3604-416f-a59a-1bcdf774f102",
   "metadata": {},
   "outputs": [],
   "source": [
    "p_meta_file = 'pseudochromosome_metadata.tsv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ad3d97e3-df32-4212-9acd-8ecd6d6069b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get a table that matches genotype + pseudochrom + human gene + mouse gene\n",
    "temp = df[['genotype', 'pseudochrom']].explode('pseudochrom')\n",
    "p_meta = pd.read_csv(p_meta_file, sep='\\t')\n",
    "p_meta.fillna('dummy', inplace=True)\n",
    "p_meta = temp.merge(p_meta, on='pseudochrom')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "559dbc4b-dcfa-47ab-a6a4-36fb23e5fdb5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genotype</th>\n",
       "      <th>pseudochrom</th>\n",
       "      <th>human_gene</th>\n",
       "      <th>mouse_gene</th>\n",
       "      <th>locus_type</th>\n",
       "      <th>notes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hTREM2_KI_HET</td>\n",
       "      <td>hTREM2</td>\n",
       "      <td>TREM2</td>\n",
       "      <td>dummy</td>\n",
       "      <td>chimeric</td>\n",
       "      <td>dummy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hTREM2_KI_WT</td>\n",
       "      <td>hTREM2</td>\n",
       "      <td>TREM2</td>\n",
       "      <td>dummy</td>\n",
       "      <td>chimeric</td>\n",
       "      <td>dummy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hTREM2_KI_WT</td>\n",
       "      <td>hTREM2</td>\n",
       "      <td>TREM2</td>\n",
       "      <td>dummy</td>\n",
       "      <td>chimeric</td>\n",
       "      <td>dummy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hgenotype1_genotype2</td>\n",
       "      <td>hgenotype1</td>\n",
       "      <td>dummy</td>\n",
       "      <td>dummy</td>\n",
       "      <td>dummy</td>\n",
       "      <td>dummy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hgenotype1_genotype2</td>\n",
       "      <td>hgenotype2</td>\n",
       "      <td>dummy</td>\n",
       "      <td>dummy</td>\n",
       "      <td>dummy</td>\n",
       "      <td>dummy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               genotype pseudochrom human_gene mouse_gene locus_type  notes\n",
       "0         hTREM2_KI_HET      hTREM2      TREM2      dummy   chimeric  dummy\n",
       "1          hTREM2_KI_WT      hTREM2      TREM2      dummy   chimeric  dummy\n",
       "2          hTREM2_KI_WT      hTREM2      TREM2      dummy   chimeric  dummy\n",
       "3  hgenotype1_genotype2  hgenotype1      dummy      dummy      dummy  dummy\n",
       "4  hgenotype1_genotype2  hgenotype2      dummy      dummy      dummy  dummy"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_meta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c113440a-a3d5-4a41-a245-675a8e50c5cd",
   "metadata": {},
   "source": [
    "## 231102 debug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03ac88da-90b1-4444-8193-d964c684c194",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> \u001b[0;32m/Users/fairliereese/Documents/programming/mortazavi_lab/bin/modelad_pipeline/proc_update/sm_utils.py\u001b[0m(164)\u001b[0;36mparse_config_file\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    162 \u001b[0;31m    \u001b[0;31m# but first sort on study and sample such that they will always be ordered in the same way\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    163 \u001b[0;31m    \u001b[0;31m# this should freeze our results\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m--> 164 \u001b[0;31m    \u001b[0mgb_cols\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m'study'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'genotype'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'sex'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'age'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'tissue'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    165 \u001b[0;31m    \u001b[0mdf\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mby\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgb_cols\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mascending\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    166 \u001b[0;31m    \u001b[0mtemp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdeep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "ipdb>  df.head()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    batch platform                                              fname  \\\n",
      "0  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
      "1  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
      "2  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
      "3  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
      "4  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
      "\n",
      "                            basename  \\\n",
      "0     ad003_12670_lig-blk_1_t1.fastq   \n",
      "1     ad003_12670_lig-blk_2_t1.fastq   \n",
      "2  ad004_18503_lig-blk_1_t1.fastq.gz   \n",
      "3  ad004_18503_lig-blk_2_t1.fastq.gz   \n",
      "4  ad004_18634_lig-blk_1_t1.fastq.gz   \n",
      "\n",
      "                                                path  flowcell  \\\n",
      "0  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...         1   \n",
      "1  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...         2   \n",
      "2  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...         1   \n",
      "3  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...         2   \n",
      "4  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...         1   \n",
      "\n",
      "               file_stem  chop_num  sample_temp  mouse_id  ...       age  \\\n",
      "0  ad003_12670_lig-blk_1       1.0  ad003_12670     12670  ...  4_months   \n",
      "1  ad003_12670_lig-blk_2       1.0  ad003_12670     12670  ...  4_months   \n",
      "2  ad004_18503_lig-blk_1       1.0  ad004_18503     18503  ...   8_weeks   \n",
      "3  ad004_18503_lig-blk_2       1.0  ad004_18503     18503  ...   8_weeks   \n",
      "4  ad004_18634_lig-blk_1       1.0  ad004_18634     18634  ...   8_weeks   \n",
      "\n",
      "  tissue pseudochrom_needed pseudochrom human_gene  mouse_gene  \\\n",
      "0     HC               True     (hCLU,)     (CLU,)      (Clu,)   \n",
      "1     HC               True     (hCLU,)     (CLU,)      (Clu,)   \n",
      "2     HC               True   (hTREM2,)   (TREM2,)    (dummy,)   \n",
      "3     HC               True   (hTREM2,)   (TREM2,)    (dummy,)   \n",
      "4     HC               True   (hTREM2,)   (TREM2,)    (dummy,)   \n",
      "\n",
      "                        sample biorep_num                        dataset  \\\n",
      "0  CLU-h2kbKI_HO_F_4_months_HC          1  CLU-h2kbKI_HO_F_4_months_HC_1   \n",
      "1  CLU-h2kbKI_HO_F_4_months_HC          1  CLU-h2kbKI_HO_F_4_months_HC_1   \n",
      "2    hTREM2_KI_WT_M_8_weeks_HC          1    hTREM2_KI_WT_M_8_weeks_HC_1   \n",
      "3    hTREM2_KI_WT_M_8_weeks_HC          1    hTREM2_KI_WT_M_8_weeks_HC_1   \n",
      "4   hTREM2_KI_HET_M_8_weeks_HC          1   hTREM2_KI_HET_M_8_weeks_HC_1   \n",
      "\n",
      "                              source  \n",
      "0  ad003_CLU-h2kbKI_HO_F_4_months_HC  \n",
      "1  ad003_CLU-h2kbKI_HO_F_4_months_HC  \n",
      "2    ad004_hTREM2_KI_WT_M_8_weeks_HC  \n",
      "3    ad004_hTREM2_KI_WT_M_8_weeks_HC  \n",
      "4   ad004_hTREM2_KI_HET_M_8_weeks_HC  \n",
      "\n",
      "[5 rows x 23 columns]\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "ipdb>  df.loc[df.genotype=='hgenotype1_genotype2']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    batch platform                                  fname  \\\n",
      "5  230516      ONT  /test/ad004_666_lig-blk_1_t1.fastq.gz   \n",
      "\n",
      "                          basename   path  flowcell            file_stem  \\\n",
      "5  ad004_666_lig-blk_1_t1.fastq.gz  /test         1  ad004_666_lig-blk_1   \n",
      "\n",
      "   chop_num sample_temp  mouse_id  ...      age tissue pseudochrom_needed  \\\n",
      "5       1.0   ad004_666       666  ...  8_weeks     HC               True   \n",
      "\n",
      "                pseudochrom        human_gene        mouse_gene  \\\n",
      "5  (hgenotype1, hgenotype2)  (hgene1, hgene2)  (hgene1, mgene2)   \n",
      "\n",
      "                              sample biorep_num  \\\n",
      "5  hgenotype1_genotype2_M_8_weeks_HC          1   \n",
      "\n",
      "                               dataset  \\\n",
      "5  hgenotype1_genotype2_M_8_weeks_HC_1   \n",
      "\n",
      "                                    source  \n",
      "5  ad004_hgenotype1_genotype2_M_8_weeks_HC  \n",
      "\n",
      "[1 rows x 23 columns]\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "ipdb>  q\n"
     ]
    }
   ],
   "source": [
    "# config_tsv = '231020_config.tsv'\n",
    "config_tsv = 'test_pseudochrom_parsing_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a9636d2-9169-4485-97eb-4bf33a20f1f5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "035fd53b-8241-441c-af13-b0e4bfbdc2b3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "073722ae-34d7-48f3-9871-d67094b3b2a2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c925a5c0-7940-47da-89fc-72ccafd3e982",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b03e1cb-f3f4-4924-931c-e87cec3f5e65",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "072320fc-71fc-4222-a1fc-625920c06026",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1a6265ce-7859-49ed-804b-264366db521d",
   "metadata": {},
   "source": [
    "## 231102 deubg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8768cd7d-68f4-488b-b4e1-4b4ed7c12ef3",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/Documents/programming/mortazavi_lab/bin/modelad_pipeline/proc_update/sm_utils.py:126: UserWarning: Config found non-pseudochrom mouse w/ genotypes ['hABKI_SWE'], is this expected?\n",
      "  warnings.warn(f'Config found non-pseudochrom mouse w/ genotypes {genotypes}, is this expected?')\n"
     ]
    }
   ],
   "source": [
    "# config_tsv = '231020_config.tsv'\n",
    "config_tsv = '231020_config.tsv'\n",
    "# config_tsv = 'test_pseudochrom_parsing_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe,\n",
    "                       include_pseudochrom=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "91f79b69-ab4c-4e9a-b0af-4244cb5d2293",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>...</th>\n",
       "      <th>tissue</th>\n",
       "      <th>pseudochrom_needed</th>\n",
       "      <th>pseudochrom</th>\n",
       "      <th>human_gene</th>\n",
       "      <th>mouse_gene</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "      <th>source</th>\n",
       "      <th>cerberus_run</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad002_10725_lig-blk_2_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad002_10725_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad002_10725</td>\n",
       "      <td>10725</td>\n",
       "      <td>...</td>\n",
       "      <td>CTX</td>\n",
       "      <td>False</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>5xFADHEMI_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_M_18_months_CTX_1</td>\n",
       "      <td>ad002_5xFADHEMI_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad002_10725_lig-ss_blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad002_10725_lig-ss_blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad002_10725</td>\n",
       "      <td>10725</td>\n",
       "      <td>...</td>\n",
       "      <td>CTX</td>\n",
       "      <td>False</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>5xFADHEMI_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_M_18_months_CTX_1</td>\n",
       "      <td>ad002_5xFADHEMI_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad002_10721_lig-blk_2_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad002_10721_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad002_10721</td>\n",
       "      <td>10721</td>\n",
       "      <td>...</td>\n",
       "      <td>CTX</td>\n",
       "      <td>False</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>5xFADWT_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADWT_M_18_months_CTX_1</td>\n",
       "      <td>ad002_5xFADWT_M_18_months_CTX</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad002_10721_lig-ss_blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad002_10721_lig-ss_blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad002_10721</td>\n",
       "      <td>10721</td>\n",
       "      <td>...</td>\n",
       "      <td>CTX</td>\n",
       "      <td>False</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>5xFADWT_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADWT_M_18_months_CTX_1</td>\n",
       "      <td>ad002_5xFADWT_M_18_months_CTX</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12517_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_12517_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12517</td>\n",
       "      <td>12517</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>True</td>\n",
       "      <td>(hCLU,)</td>\n",
       "      <td>(CLU,)</td>\n",
       "      <td>(Clu,)</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC_1</td>\n",
       "      <td>ad003_5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    batch platform                                              fname  \\\n",
       "0  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "1  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "2  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "3  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "4  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                               basename  \\\n",
       "0     ad002_10725_lig-blk_2_t1.fastq.gz   \n",
       "1  ad002_10725_lig-ss_blk_1_t1.fastq.gz   \n",
       "2     ad002_10721_lig-blk_2_t1.fastq.gz   \n",
       "3  ad002_10721_lig-ss_blk_1_t1.fastq.gz   \n",
       "4        ad003_12517_lig-blk_1_t1.fastq   \n",
       "\n",
       "                                                path flowcell  \\\n",
       "0  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "1  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "2  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "3  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "4  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "\n",
       "                  file_stem  chop_num  sample_temp  mouse_id  ... tissue  \\\n",
       "0     ad002_10725_lig-blk_2       1.0  ad002_10725     10725  ...    CTX   \n",
       "1  ad002_10725_lig-ss_blk_1       1.0  ad002_10725     10725  ...    CTX   \n",
       "2     ad002_10721_lig-blk_2       1.0  ad002_10721     10721  ...    CTX   \n",
       "3  ad002_10721_lig-ss_blk_1       1.0  ad002_10721     10721  ...    CTX   \n",
       "4     ad003_12517_lig-blk_1       1.0  ad003_12517     12517  ...     HC   \n",
       "\n",
       "  pseudochrom_needed pseudochrom human_gene mouse_gene  \\\n",
       "0              False    (dummy,)   (dummy,)   (dummy,)   \n",
       "1              False    (dummy,)   (dummy,)   (dummy,)   \n",
       "2              False    (dummy,)   (dummy,)   (dummy,)   \n",
       "3              False    (dummy,)   (dummy,)   (dummy,)   \n",
       "4               True     (hCLU,)     (CLU,)     (Clu,)   \n",
       "\n",
       "                          sample biorep_num                          dataset  \\\n",
       "0      5xFADHEMI_M_18_months_CTX          1      5xFADHEMI_M_18_months_CTX_1   \n",
       "1      5xFADHEMI_M_18_months_CTX          1      5xFADHEMI_M_18_months_CTX_1   \n",
       "2        5xFADWT_M_18_months_CTX          1        5xFADWT_M_18_months_CTX_1   \n",
       "3        5xFADWT_M_18_months_CTX          1        5xFADWT_M_18_months_CTX_1   \n",
       "4  5xCLU-h2kbKI_HO_F_4_months_HC          1  5xCLU-h2kbKI_HO_F_4_months_HC_1   \n",
       "\n",
       "                                source cerberus_run  \n",
       "0      ad002_5xFADHEMI_M_18_months_CTX            1  \n",
       "1      ad002_5xFADHEMI_M_18_months_CTX            1  \n",
       "2        ad002_5xFADWT_M_18_months_CTX            2  \n",
       "3        ad002_5xFADWT_M_18_months_CTX            2  \n",
       "4  ad003_5xCLU-h2kbKI_HO_F_4_months_HC            3  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e66b749e-ef86-4843-9420-b4fc977bb18a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['5xFADHEMI', '5xFADWT', '5xCLU-h2kbKI_HO', 'CLU-h2kbKI_HO',\n",
       "       'hTREM2-R47HKI_HET', 'hTREM2_KI_HET', 'hTREM2_KI_WT', '5xBIN1_HO',\n",
       "       'BIN1_HO', 'C57B6J', 'hABKI_SWE', 'ABCA7_HO'], dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.genotype.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b7532338-4401-4121-8c64-7cca7d585882",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>...</th>\n",
       "      <th>tissue</th>\n",
       "      <th>pseudochrom_needed</th>\n",
       "      <th>pseudochrom</th>\n",
       "      <th>human_gene</th>\n",
       "      <th>mouse_gene</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "      <th>source</th>\n",
       "      <th>cerberus_run</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad002_10725_lig-blk_2_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad002_10725_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad002_10725</td>\n",
       "      <td>10725</td>\n",
       "      <td>...</td>\n",
       "      <td>CTX</td>\n",
       "      <td>False</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>5xFADHEMI_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_M_18_months_CTX_1</td>\n",
       "      <td>ad002_5xFADHEMI_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad002_10725_lig-ss_blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad002_10725_lig-ss_blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad002_10725</td>\n",
       "      <td>10725</td>\n",
       "      <td>...</td>\n",
       "      <td>CTX</td>\n",
       "      <td>False</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>5xFADHEMI_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_M_18_months_CTX_1</td>\n",
       "      <td>ad002_5xFADHEMI_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11616_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "      <td>ad003_5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11616_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "      <td>ad003_5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11617_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11617_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11617</td>\n",
       "      <td>11617</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "      <td>ad003_5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11617_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11617_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11617</td>\n",
       "      <td>11617</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "      <td>ad003_5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11625_lig-blk_1_t2.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11625_lig-blk_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ad003_11625</td>\n",
       "      <td>11625</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_3</td>\n",
       "      <td>ad003_5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11625_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11625_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11625</td>\n",
       "      <td>11625</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>(dummy,)</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_3</td>\n",
       "      <td>ad003_5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     batch platform                                              fname  \\\n",
       "0   230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "1   230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "10  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "11  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "12  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "13  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "14  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "15  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                                basename  \\\n",
       "0      ad002_10725_lig-blk_2_t1.fastq.gz   \n",
       "1   ad002_10725_lig-ss_blk_1_t1.fastq.gz   \n",
       "10        ad003_11616_lig-blk_1_t1.fastq   \n",
       "11        ad003_11616_lig-blk_2_t1.fastq   \n",
       "12        ad003_11617_lig-blk_1_t1.fastq   \n",
       "13        ad003_11617_lig-blk_2_t1.fastq   \n",
       "14        ad003_11625_lig-blk_1_t2.fastq   \n",
       "15        ad003_11625_lig-blk_2_t1.fastq   \n",
       "\n",
       "                                                 path flowcell  \\\n",
       "0   /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "1   /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "10  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "11  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "12  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "13  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "14  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "15  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "\n",
       "                   file_stem  chop_num  sample_temp  mouse_id  ... tissue  \\\n",
       "0      ad002_10725_lig-blk_2       1.0  ad002_10725     10725  ...    CTX   \n",
       "1   ad002_10725_lig-ss_blk_1       1.0  ad002_10725     10725  ...    CTX   \n",
       "10     ad003_11616_lig-blk_1       1.0  ad003_11616     11616  ...     HC   \n",
       "11     ad003_11616_lig-blk_2       1.0  ad003_11616     11616  ...     HC   \n",
       "12     ad003_11617_lig-blk_1       1.0  ad003_11617     11617  ...     HC   \n",
       "13     ad003_11617_lig-blk_2       1.0  ad003_11617     11617  ...     HC   \n",
       "14     ad003_11625_lig-blk_1       2.0  ad003_11625     11625  ...     HC   \n",
       "15     ad003_11625_lig-blk_2       1.0  ad003_11625     11625  ...     HC   \n",
       "\n",
       "   pseudochrom_needed pseudochrom human_gene mouse_gene  \\\n",
       "0               False    (dummy,)   (dummy,)   (dummy,)   \n",
       "1               False    (dummy,)   (dummy,)   (dummy,)   \n",
       "10              False    (dummy,)   (dummy,)   (dummy,)   \n",
       "11              False    (dummy,)   (dummy,)   (dummy,)   \n",
       "12              False    (dummy,)   (dummy,)   (dummy,)   \n",
       "13              False    (dummy,)   (dummy,)   (dummy,)   \n",
       "14              False    (dummy,)   (dummy,)   (dummy,)   \n",
       "15              False    (dummy,)   (dummy,)   (dummy,)   \n",
       "\n",
       "                       sample biorep_num                      dataset  \\\n",
       "0   5xFADHEMI_M_18_months_CTX          1  5xFADHEMI_M_18_months_CTX_1   \n",
       "1   5xFADHEMI_M_18_months_CTX          1  5xFADHEMI_M_18_months_CTX_1   \n",
       "10    5xFADHEMI_F_4_months_HC          1    5xFADHEMI_F_4_months_HC_1   \n",
       "11    5xFADHEMI_F_4_months_HC          1    5xFADHEMI_F_4_months_HC_1   \n",
       "12    5xFADHEMI_F_4_months_HC          2    5xFADHEMI_F_4_months_HC_2   \n",
       "13    5xFADHEMI_F_4_months_HC          2    5xFADHEMI_F_4_months_HC_2   \n",
       "14    5xFADHEMI_F_4_months_HC          3    5xFADHEMI_F_4_months_HC_3   \n",
       "15    5xFADHEMI_F_4_months_HC          3    5xFADHEMI_F_4_months_HC_3   \n",
       "\n",
       "                             source cerberus_run  \n",
       "0   ad002_5xFADHEMI_M_18_months_CTX            1  \n",
       "1   ad002_5xFADHEMI_M_18_months_CTX            1  \n",
       "10    ad003_5xFADHEMI_F_4_months_HC            4  \n",
       "11    ad003_5xFADHEMI_F_4_months_HC            4  \n",
       "12    ad003_5xFADHEMI_F_4_months_HC            4  \n",
       "13    ad003_5xFADHEMI_F_4_months_HC            4  \n",
       "14    ad003_5xFADHEMI_F_4_months_HC            4  \n",
       "15    ad003_5xFADHEMI_F_4_months_HC            4  \n",
       "\n",
       "[8 rows x 24 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.genotype=='5xFADHEMI']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "65e333ac-e03d-4346-b132-f58f1982bf43",
   "metadata": {},
   "outputs": [],
   "source": [
    "#   study=ad002\n",
    "\n",
    "#   genotype=5xFADHEMI\n",
    "#   sex=M\n",
    "#   age=18_months\n",
    "#   tissue=CTX\n",
    "wc = {'study': 'ad002', 'genotype': '5xFADHEMI',\n",
    "      'sex': 'M', 'age': '18_months', 'tissue': 'CTX'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a79a4007-d51d-42a4-b7d8-37818e91d793",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/merge/ad002/5xFADHEMI_M_18_months_CTX_1_sorted.bam']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_cfg_entries(wc, df, config['merge']['sort_bam'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae68aae8-a258-4df2-bdca-41d00c898f35",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "964376b2-f333-4fd8-91c4-811d1745c2e1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dab8207-1d7d-49fb-b2dc-c85e6b612deb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "437aebc5-90d5-4038-9c7b-cda96a193b7d",
   "metadata": {},
   "source": [
    "## 231102 adding hgene / mgene parsing to config parser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d3860a9d-7680-466a-829c-1895f0661d30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# config_tsv = '231020_config.tsv'\n",
    "config_tsv = 'test_pseudochrom_parsing_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe,\n",
    "                       include_pseudochrom=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "42463b97-fac8-42d3-8a5b-0077696f4389",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>...</th>\n",
       "      <th>tissue</th>\n",
       "      <th>pseudochrom_needed</th>\n",
       "      <th>pseudochrom</th>\n",
       "      <th>human_gene</th>\n",
       "      <th>mouse_gene</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "      <th>source</th>\n",
       "      <th>cerberus_run</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/test/ad004_666_lig-blk_1_t1.fastq.gz</td>\n",
       "      <td>ad004_666_lig-blk_1_t1.fastq.gz</td>\n",
       "      <td>/test</td>\n",
       "      <td>1</td>\n",
       "      <td>ad004_666_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad004_666</td>\n",
       "      <td>666</td>\n",
       "      <td>...</td>\n",
       "      <td>HC</td>\n",
       "      <td>True</td>\n",
       "      <td>(hgenotype1, hgenotype2)</td>\n",
       "      <td>(hgene1, hgene2)</td>\n",
       "      <td>(hgene1, mgene2)</td>\n",
       "      <td>hgenotype1_genotype2_M_8_weeks_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>hgenotype1_genotype2_M_8_weeks_HC_1</td>\n",
       "      <td>ad004_hgenotype1_genotype2_M_8_weeks_HC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    batch platform                                  fname  \\\n",
       "5  230516      ONT  /test/ad004_666_lig-blk_1_t1.fastq.gz   \n",
       "\n",
       "                          basename   path flowcell            file_stem  \\\n",
       "5  ad004_666_lig-blk_1_t1.fastq.gz  /test        1  ad004_666_lig-blk_1   \n",
       "\n",
       "   chop_num sample_temp  mouse_id  ... tissue pseudochrom_needed  \\\n",
       "5       1.0   ad004_666       666  ...     HC               True   \n",
       "\n",
       "                pseudochrom        human_gene        mouse_gene  \\\n",
       "5  (hgenotype1, hgenotype2)  (hgene1, hgene2)  (hgene1, mgene2)   \n",
       "\n",
       "                              sample biorep_num  \\\n",
       "5  hgenotype1_genotype2_M_8_weeks_HC          1   \n",
       "\n",
       "                               dataset  \\\n",
       "5  hgenotype1_genotype2_M_8_weeks_HC_1   \n",
       "\n",
       "                                    source cerberus_run  \n",
       "5  ad004_hgenotype1_genotype2_M_8_weeks_HC            4  \n",
       "\n",
       "[1 rows x 24 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.genotype=='hgenotype1_genotype2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5049a39b-513a-4706-a06d-303cb50c5237",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df['pseudochromosome_names'] = df.apply(lambda x: tuple(sorted(x.pseudochromosome_names)), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2be4f549-5900-47ae-8858-645e145adc42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.loc[df.genotype=='hgenotype1_genotype2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a9c649c6-1f28-49e9-a9bd-d6bb37140b36",
   "metadata": {},
   "outputs": [],
   "source": [
    "wc = {'pseudochrom': ('hgenotype1', 'hgenotype2')}\n",
    "# get_df_col(wc, df, 'genotype')\n",
    "# get_cfg_entries(wc, \n",
    "#                 df,\n",
    "#                 config['ref']['pseudochrom']['fa_merge'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e99d84ca-2994-47c4-bbb3-f58a68dd88a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ref/pseudochrom/hgenotype1/ref.fa', 'ref/pseudochrom/hgenotype2/ref.fa']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get all the pseudo chromosomes that are needed \n",
    "# need for the given genotype\n",
    "get_cfg_entries(wc, \n",
    "     df,\n",
    "     config['ref']['pseudochrom']['fa'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "e1a2ccf1-ca79-45e0-8390-8e1e86a88c6e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ref/genome/hgenotype1_genotype2/merged.fa']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# output\n",
    "get_cfg_entries(wc, \n",
    "                df,\n",
    "                config['ref']['pseudochrom']['fa_merge'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a68064c8-6e37-4d25-9e68-54c2d6d036c7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67ac335e-451c-4f62-9b3b-9b7f3d109eff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "2b800bd7-9ab0-479a-bddb-42b52537b52f",
   "metadata": {},
   "source": [
    "## 231023 -- get the cerberus run = 1 for the first agg runs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "db8ee3d5-78f1-4255-8cf4-1d0f033c1c1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_tsv = '231020_config.tsv'\n",
    "\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "# meta_tsv = 'mouse_test_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe,\n",
    "                       include_humanized=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e23aa3c1-ecd3-4d4d-bacd-c1d24077b073",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>...</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>tissue</th>\n",
       "      <th>humanized</th>\n",
       "      <th>humanized_loci</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "      <th>source</th>\n",
       "      <th>cerberus_run</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad002_10725_lig-blk_2_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad002_10725_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad002_10725</td>\n",
       "      <td>10725</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>18_months</td>\n",
       "      <td>CTX</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5xFADHEMI_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_M_18_months_CTX_1</td>\n",
       "      <td>ad002_5xFADHEMI_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad002_10725_lig-ss_blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad002_10725_lig-ss_blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad002_10725</td>\n",
       "      <td>10725</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>18_months</td>\n",
       "      <td>CTX</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5xFADHEMI_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_M_18_months_CTX_1</td>\n",
       "      <td>ad002_5xFADHEMI_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad002_10721_lig-blk_2_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad002_10721_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad002_10721</td>\n",
       "      <td>10721</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>18_months</td>\n",
       "      <td>CTX</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5xFADWT_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADWT_M_18_months_CTX_1</td>\n",
       "      <td>ad002_5xFADWT_M_18_months_CTX</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad002_10721_lig-ss_blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad002_10721_lig-ss_blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad002_10721</td>\n",
       "      <td>10721</td>\n",
       "      <td>...</td>\n",
       "      <td>M</td>\n",
       "      <td>18_months</td>\n",
       "      <td>CTX</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5xFADWT_M_18_months_CTX</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADWT_M_18_months_CTX_1</td>\n",
       "      <td>ad002_5xFADWT_M_18_months_CTX</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11616_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>...</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "      <td>ad003_5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    batch platform                                              fname  \\\n",
       "0  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "1  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "2  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "3  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "4  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                               basename  \\\n",
       "0     ad002_10725_lig-blk_2_t1.fastq.gz   \n",
       "1  ad002_10725_lig-ss_blk_1_t1.fastq.gz   \n",
       "2     ad002_10721_lig-blk_2_t1.fastq.gz   \n",
       "3  ad002_10721_lig-ss_blk_1_t1.fastq.gz   \n",
       "4        ad003_11616_lig-blk_1_t1.fastq   \n",
       "\n",
       "                                                path flowcell  \\\n",
       "0  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "1  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "2  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "3  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "4  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "\n",
       "                  file_stem  chop_num  sample_temp  mouse_id  ... sex  \\\n",
       "0     ad002_10725_lig-blk_2       1.0  ad002_10725     10725  ...   M   \n",
       "1  ad002_10725_lig-ss_blk_1       1.0  ad002_10725     10725  ...   M   \n",
       "2     ad002_10721_lig-blk_2       1.0  ad002_10721     10721  ...   M   \n",
       "3  ad002_10721_lig-ss_blk_1       1.0  ad002_10721     10721  ...   M   \n",
       "4     ad003_11616_lig-blk_1       1.0  ad003_11616     11616  ...   F   \n",
       "\n",
       "         age tissue humanized humanized_loci                     sample  \\\n",
       "0  18_months    CTX     False            NaN  5xFADHEMI_M_18_months_CTX   \n",
       "1  18_months    CTX     False            NaN  5xFADHEMI_M_18_months_CTX   \n",
       "2  18_months    CTX     False            NaN    5xFADWT_M_18_months_CTX   \n",
       "3  18_months    CTX     False            NaN    5xFADWT_M_18_months_CTX   \n",
       "4   4_months     HC     False            NaN    5xFADHEMI_F_4_months_HC   \n",
       "\n",
       "   biorep_num                      dataset                           source  \\\n",
       "0           1  5xFADHEMI_M_18_months_CTX_1  ad002_5xFADHEMI_M_18_months_CTX   \n",
       "1           1  5xFADHEMI_M_18_months_CTX_1  ad002_5xFADHEMI_M_18_months_CTX   \n",
       "2           1    5xFADWT_M_18_months_CTX_1    ad002_5xFADWT_M_18_months_CTX   \n",
       "3           1    5xFADWT_M_18_months_CTX_1    ad002_5xFADWT_M_18_months_CTX   \n",
       "4           1    5xFADHEMI_F_4_months_HC_1    ad003_5xFADHEMI_F_4_months_HC   \n",
       "\n",
       "  cerberus_run  \n",
       "0            1  \n",
       "1            1  \n",
       "2            2  \n",
       "3            2  \n",
       "4            3  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e9ca2ce1-c415-4200-9e17-93aa7a5dc722",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/cerberus/ad002/5xFADHEMI_M_18_months_CTX_1_ic.tsv'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cfg_entry = config['cerberus']['ics']\n",
    "wc = {}\n",
    "get_first_cerb_entry(wc, df, cfg_entry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7461e0b8-3b2b-4b38-b42a-cebc5d4992eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/cerberus/ad002/5xFADHEMI_M_18_months_CTX_1_tss.bed'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cfg_entry = config['cerberus']['ends']\n",
    "wc = {'end_mode': 'tss'}\n",
    "get_first_cerb_entry(wc, df, cfg_entry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3714c71e-1300-452c-b28b-841d4c534b58",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9be2f30-faae-4670-ad6a-55b1c38f9ad0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c61521c5-7c1f-487e-bfd7-493b30305a7c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1d56570-9f23-47cd-a1d4-e5539c4a8f40",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f30b0610-5680-4533-b4ee-7f1d4d115c72",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "62881d45-904f-4f5e-b698-f8009c15de4d",
   "metadata": {},
   "source": [
    "## 231022 -- some other suff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "fbd49bff-fc8a-494f-9cfa-814351a3501b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/merge/ad005/5xBIN1_HO_F_4_months_HC_3_sorted.bam',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_2_sorted.bam',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_1_sorted.bam']"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wc = {'study': 'ad005',\n",
    "      'genotype': '5xBIN1_HO',\n",
    "      'sex': 'F',\n",
    "      'age': '4_months',\n",
    "      'tissue': 'HC'}\n",
    "\n",
    "config_tsv = '231020_config.tsv'\n",
    "# config_tsv = 'test_local_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "# meta_tsv = 'mouse_test_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe,\n",
    "                       include_humanized=False)\n",
    "get_cfg_entries(wc, df, config['merge']['sort_bam'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "172d39b6-2b45-4eae-9d5b-d3212661e9ff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac458e1c-b91c-404c-b4f2-0c9a47002fd7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4efdbc27-72d8-4d34-b152-f3e5a8cc4feb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41ec23b3-b14c-481a-afb6-ec2f67163b26",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92f0ea7e-ab19-4c21-8470-94e65cec1a3c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87716ce6-d698-4ebc-8ee7-e97598121bfc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d70a1a4-e05d-48c1-842b-38550f373e7f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d9dd5d75-ac5b-478a-99d7-8d39b1a9bd00",
   "metadata": {},
   "source": [
    "## 231020 pipeline restructuring -- get bws for each mouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "48824966-4dea-4232-854e-4d3acadd375e",
   "metadata": {},
   "outputs": [],
   "source": [
    "wc = {'study': 'ad003',\n",
    "      'genotype': '5xFADHEMI',\n",
    "      'sex': 'F',\n",
    "      'age': '4_months',\n",
    "      'tissue': 'HC'}\n",
    "\n",
    "config_tsv = '231020_config.tsv'\n",
    "# config_tsv = 'test_local_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "# meta_tsv = 'mouse_test_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe,\n",
    "                       include_humanized=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5f9c7db4-8c85-4eac-ac21-b6ee353b584b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# expand(get_cfg_entries(wc, df,\n",
    "#                         config['merge']['bw']),\n",
    "#        strand=['fwd', 'rev'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d5b9ffcb-d1eb-4ad3-adab-645003263f0b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/merge/ad002/5xFADHEMI_M_18_months_CTX_1_fwd.bw',\n",
       " 'data/merge/ad002/5xFADHEMI_M_18_months_CTX_1_rev.bw',\n",
       " 'data/merge/ad002/5xFADHEMI_M_18_months_CTX_1_fwd.bw',\n",
       " 'data/merge/ad002/5xFADHEMI_M_18_months_CTX_1_rev.bw',\n",
       " 'data/merge/ad002/5xFADWT_M_18_months_CTX_1_fwd.bw',\n",
       " 'data/merge/ad002/5xFADWT_M_18_months_CTX_1_rev.bw',\n",
       " 'data/merge/ad002/5xFADWT_M_18_months_CTX_1_fwd.bw',\n",
       " 'data/merge/ad002/5xFADWT_M_18_months_CTX_1_rev.bw',\n",
       " 'data/merge/ad003/5xFADHEMI_F_4_months_HC_1_fwd.bw',\n",
       " 'data/merge/ad003/5xFADHEMI_F_4_months_HC_1_rev.bw',\n",
       " 'data/merge/ad003/5xFADHEMI_F_4_months_HC_1_fwd.bw',\n",
       " 'data/merge/ad003/5xFADHEMI_F_4_months_HC_1_rev.bw',\n",
       " 'data/merge/ad003/5xFADHEMI_F_4_months_HC_2_fwd.bw',\n",
       " 'data/merge/ad003/5xFADHEMI_F_4_months_HC_2_rev.bw',\n",
       " 'data/merge/ad003/5xFADHEMI_F_4_months_HC_2_fwd.bw',\n",
       " 'data/merge/ad003/5xFADHEMI_F_4_months_HC_2_rev.bw',\n",
       " 'data/merge/ad003/5xFADHEMI_F_4_months_HC_3_fwd.bw',\n",
       " 'data/merge/ad003/5xFADHEMI_F_4_months_HC_3_rev.bw',\n",
       " 'data/merge/ad003/5xFADHEMI_F_4_months_HC_3_fwd.bw',\n",
       " 'data/merge/ad003/5xFADHEMI_F_4_months_HC_3_rev.bw',\n",
       " 'data/merge/ad003/5xFADWT_F_4_months_HC_1_fwd.bw',\n",
       " 'data/merge/ad003/5xFADWT_F_4_months_HC_1_rev.bw',\n",
       " 'data/merge/ad003/5xFADWT_F_4_months_HC_1_fwd.bw',\n",
       " 'data/merge/ad003/5xFADWT_F_4_months_HC_1_rev.bw',\n",
       " 'data/merge/ad003/5xFADWT_F_4_months_HC_2_fwd.bw',\n",
       " 'data/merge/ad003/5xFADWT_F_4_months_HC_2_rev.bw',\n",
       " 'data/merge/ad003/5xFADWT_F_4_months_HC_2_fwd.bw',\n",
       " 'data/merge/ad003/5xFADWT_F_4_months_HC_2_rev.bw',\n",
       " 'data/merge/ad003/5xFADWT_F_4_months_HC_3_fwd.bw',\n",
       " 'data/merge/ad003/5xFADWT_F_4_months_HC_3_rev.bw',\n",
       " 'data/merge/ad003/5xFADWT_F_4_months_HC_3_fwd.bw',\n",
       " 'data/merge/ad003/5xFADWT_F_4_months_HC_3_rev.bw',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_3_fwd.bw',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_3_rev.bw',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_2_fwd.bw',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_2_rev.bw',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_3_fwd.bw',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_3_rev.bw',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_1_fwd.bw',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_1_rev.bw',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_2_fwd.bw',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_2_rev.bw',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_1_fwd.bw',\n",
       " 'data/merge/ad005/5xBIN1_HO_F_4_months_HC_1_rev.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_3_fwd.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_3_rev.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_1_fwd.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_1_rev.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_2_fwd.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_2_rev.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_3_fwd.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_3_rev.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_2_fwd.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_2_rev.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_2_fwd.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_2_rev.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_1_fwd.bw',\n",
       " 'data/merge/ad005/BIN1_HO_F_4_months_HC_1_rev.bw',\n",
       " 'data/merge/ad006/C57B6J_F_8_weeks_HC_3_fwd.bw',\n",
       " 'data/merge/ad006/C57B6J_F_8_weeks_HC_3_rev.bw',\n",
       " 'data/merge/ad006/C57B6J_F_8_weeks_HC_1_fwd.bw',\n",
       " 'data/merge/ad006/C57B6J_F_8_weeks_HC_1_rev.bw',\n",
       " 'data/merge/ad006/C57B6J_F_8_weeks_HC_1_fwd.bw',\n",
       " 'data/merge/ad006/C57B6J_F_8_weeks_HC_1_rev.bw',\n",
       " 'data/merge/ad006/C57B6J_F_8_weeks_HC_2_fwd.bw',\n",
       " 'data/merge/ad006/C57B6J_F_8_weeks_HC_2_rev.bw',\n",
       " 'data/merge/ad006/C57B6J_F_8_weeks_HC_3_fwd.bw',\n",
       " 'data/merge/ad006/C57B6J_F_8_weeks_HC_3_rev.bw',\n",
       " 'data/merge/ad006/C57B6J_F_8_weeks_HC_2_fwd.bw',\n",
       " 'data/merge/ad006/C57B6J_F_8_weeks_HC_2_rev.bw',\n",
       " 'data/merge/ad006/C57B6J_M_8_weeks_HC_1_fwd.bw',\n",
       " 'data/merge/ad006/C57B6J_M_8_weeks_HC_1_rev.bw',\n",
       " 'data/merge/ad006/C57B6J_M_8_weeks_HC_3_fwd.bw',\n",
       " 'data/merge/ad006/C57B6J_M_8_weeks_HC_3_rev.bw',\n",
       " 'data/merge/ad006/C57B6J_M_8_weeks_HC_1_fwd.bw',\n",
       " 'data/merge/ad006/C57B6J_M_8_weeks_HC_1_rev.bw',\n",
       " 'data/merge/ad006/C57B6J_M_8_weeks_HC_3_fwd.bw',\n",
       " 'data/merge/ad006/C57B6J_M_8_weeks_HC_3_rev.bw',\n",
       " 'data/merge/ad006/C57B6J_M_8_weeks_HC_2_fwd.bw',\n",
       " 'data/merge/ad006/C57B6J_M_8_weeks_HC_2_rev.bw',\n",
       " 'data/merge/ad006/C57B6J_M_8_weeks_HC_2_fwd.bw',\n",
       " 'data/merge/ad006/C57B6J_M_8_weeks_HC_2_rev.bw',\n",
       " 'data/merge/ad007/ABCA7_HO_F_4_months_HC_3_fwd.bw',\n",
       " 'data/merge/ad007/ABCA7_HO_F_4_months_HC_3_rev.bw',\n",
       " 'data/merge/ad007/ABCA7_HO_F_4_months_HC_2_fwd.bw',\n",
       " 'data/merge/ad007/ABCA7_HO_F_4_months_HC_2_rev.bw',\n",
       " 'data/merge/ad007/ABCA7_HO_F_4_months_HC_3_fwd.bw',\n",
       " 'data/merge/ad007/ABCA7_HO_F_4_months_HC_3_rev.bw',\n",
       " 'data/merge/ad007/ABCA7_HO_F_4_months_HC_1_fwd.bw',\n",
       " 'data/merge/ad007/ABCA7_HO_F_4_months_HC_1_rev.bw',\n",
       " 'data/merge/ad007/ABCA7_HO_F_4_months_HC_1_fwd.bw',\n",
       " 'data/merge/ad007/ABCA7_HO_F_4_months_HC_1_rev.bw',\n",
       " 'data/merge/ad007/ABCA7_HO_F_4_months_HC_2_fwd.bw',\n",
       " 'data/merge/ad007/ABCA7_HO_F_4_months_HC_2_rev.bw']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expand(expand(config['merge']['bw'],\n",
    "        zip,\n",
    "        study=df.study.tolist(),\n",
    "        genotype=df.genotype.tolist(),\n",
    "        sex=df.sex.tolist(),\n",
    "        age=df.age.tolist(),\n",
    "        tissue=df.tissue.tolist(),\n",
    "        biorep_num=df.biorep_num.tolist(),\n",
    "        allow_missing=True),\n",
    "        strand=['fwd', 'rev'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aea0023b-e580-4435-97fd-8cde08364ecc",
   "metadata": {},
   "source": [
    "## 231020 pipeline restructuring -- sanitizing humanized input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "10610bc2-eb0c-4e24-9e15-aac04d1e7966",
   "metadata": {},
   "outputs": [],
   "source": [
    "wc = {'study': 'ad003',\n",
    "      'genotype': '5xFADHEMI',\n",
    "      'sex': 'F',\n",
    "      'age': '4_months',\n",
    "      'tissue': 'HC',\n",
    "      'cerberus_run': '2'}\n",
    "\n",
    "config_tsv = '231020_config.tsv'\n",
    "# config_tsv = 'test_local_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "# meta_tsv = 'mouse_test_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe,\n",
    "                       include_humanized=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "473b7bb4-608e-4ede-bb45-94a4cf2b7d4f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study</th>\n",
       "      <th>genotype</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>tissue</th>\n",
       "      <th>humanized</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ad002</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>M</td>\n",
       "      <td>18_months</td>\n",
       "      <td>CTX</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ad002</td>\n",
       "      <td>5xFADWT</td>\n",
       "      <td>M</td>\n",
       "      <td>18_months</td>\n",
       "      <td>CTX</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADWT</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ad005</td>\n",
       "      <td>5xBIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ad005</td>\n",
       "      <td>BIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ad006</td>\n",
       "      <td>C57B6J</td>\n",
       "      <td>F</td>\n",
       "      <td>8_weeks</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>ad006</td>\n",
       "      <td>C57B6J</td>\n",
       "      <td>M</td>\n",
       "      <td>8_weeks</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>ad007</td>\n",
       "      <td>ABCA7_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    study   genotype sex        age tissue  humanized\n",
       "0   ad002  5xFADHEMI   M  18_months    CTX      False\n",
       "2   ad002    5xFADWT   M  18_months    CTX      False\n",
       "4   ad003  5xFADHEMI   F   4_months     HC      False\n",
       "10  ad003    5xFADWT   F   4_months     HC      False\n",
       "16  ad005  5xBIN1_HO   F   4_months     HC      False\n",
       "22  ad005    BIN1_HO   F   4_months     HC      False\n",
       "29  ad006     C57B6J   F    8_weeks     HC      False\n",
       "35  ad006     C57B6J   M    8_weeks     HC      False\n",
       "41  ad007   ABCA7_HO   F   4_months     HC      False"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['study', 'genotype', 'sex', 'age', 'tissue', 'humanized']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95ceb5b6-c7f2-46fe-a675-be30e1f8eefc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99bf84cc-bedc-4b5a-8fd0-216796dce9ef",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d593edc0-3708-444e-896c-37b9da9fcc17",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a10b628-7fd8-4d9d-8067-55b6a0e5ac8a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "97c7f894-190c-4e90-81da-8fb32acb662f",
   "metadata": {},
   "source": [
    "## 231020 pipeline restructuring -- more cerberus shenannigans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "7869a4b3-273e-4966-a30a-2024d5d2f598",
   "metadata": {},
   "outputs": [],
   "source": [
    "wc = {'study': 'ad003',\n",
    "      'genotype': '5xFADHEMI',\n",
    "      'sex': 'F',\n",
    "      'age': '4_months',\n",
    "      'tissue': 'HC',\n",
    "      'cerberus_run': 2}\n",
    "wc = {'study': 'ad003',\n",
    "      'genotype': '5xFADHEMI',\n",
    "      'sex': 'F',\n",
    "      'age': '4_months',\n",
    "      'tissue': 'HC'}\n",
    "\n",
    "# config_tsv = '231018_config.tsv'\n",
    "config_tsv = 'test_local_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "721ed92e-3e16-4fcc-a712-be80779837d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/cerberus/ad003/5xFADHEMI_F_4_months_HC_1.h5',\n",
       " 'data/cerberus/ad003/5xFADWT_F_4_months_HC_2.h5',\n",
       " 'data/cerberus/ad005/BIN1_HO_F_4_months_HC_3.h5']"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_cfg_entries({'cerberus_run':df.cerberus_run.unique().tolist()},\n",
    "                        df,\n",
    "                        config['cerberus']['ca'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "0585d50d-718f-414e-a0b1-abbd3a9ce27f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to get tss / tes by adding end mode wcs\n",
    "# expand(['cerberus']['agg']['ends'],\n",
    "#                                end_mode='tss')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "b0d9d1d6-78ad-4846-989f-71a60d779410",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/cerberus/agg/ad003_5xFADHEMI_F_4_months_HC_1_tss.bed']"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expand(get_cfg_entries(wc,\n",
    "                        df,\n",
    "                        config['cerberus']['agg']['ends']),\n",
    "       end_mode='tss')[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63a06f60-0d42-4036-8e8f-d44837de599a",
   "metadata": {},
   "source": [
    "## 231019 pipeline restructuring -- cerberus getting previous cerberus run"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "1abddd32-cabb-4a60-976f-374d7ba17254",
   "metadata": {},
   "outputs": [],
   "source": [
    "wc = {'study': 'ad003',\n",
    "      'genotype': '5xFADHEMI',\n",
    "      'sex': 'F',\n",
    "      'age': '4_months',\n",
    "      'tissue': 'HC',\n",
    "      'cerberus_run': '2'}\n",
    "\n",
    "# config_tsv = '231018_config.tsv'\n",
    "config_tsv = 'test_local_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "9d6a44c4-3cc6-46cb-8033-f23b2df2b7dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3'"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.cerberus_run.max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "baa5b0af-6358-4373-a4cf-cf87932b1391",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/cerberus/agg/ad005_BIN1_HO_F_4_months_HC_3_{end_mode}.bed']"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_cfg_entries({'cerberus_run':df.cerberus_run.max()},\n",
    "                        df,\n",
    "                        config['cerberus']['agg']['ends'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "ef855e6f-3d62-49b5-8ed7-510a63c3b178",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_prev_cerb_entry(wc, df, cfg_entry):\n",
    "    \"\"\"\n",
    "    Get the previous config entry run for Cerberus. Ensure that \n",
    "    only one file meets these criteria.\n",
    "    \"\"\"\n",
    "    prev_wc = {'cerberus_run': str(int(wc['cerberus_run'])-1)}\n",
    "    file = get_cfg_entries(prev_wc, df, cfg_entry)\n",
    "    assert len(file) == 1\n",
    "    file = file[0]\n",
    "    \n",
    "    # add in the end mode if we have it\n",
    "    if 'end_mode' in wc.keys():\n",
    "        file = expand(file, \n",
    "                      zip,\n",
    "                      end_mode=wc['end_mode'])\n",
    "        assert len(file) == 1\n",
    "        file = file[0]\n",
    "    \n",
    "    return file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "1e34595c-686a-4545-9312-5dc8e25d15b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/cerberus/agg/ad003_5xFADHEMI_F_4_months_HC_1_ic.tsv'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_prev_cerb_entry(wc, df, config['cerberus']['agg']['ics'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "c10617d9-630c-49fd-a66e-67ef468a7e76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/cerberus/agg/ad003_5xFADHEMI_F_4_months_HC_1_tss.bed'"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wc = {'study': 'ad003',\n",
    "      'genotype': '5xFADHEMI',\n",
    "      'sex': 'F',\n",
    "      'age': '4_months',\n",
    "      'tissue': 'HC',\n",
    "      'end_mode': 'tss',\n",
    "      'cerberus_run': '2'}\n",
    "\n",
    "# config_tsv = '231018_config.tsv'\n",
    "config_tsv = 'test_local_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)\n",
    "\n",
    "get_prev_cerb_entry(wc, df, config['cerberus']['agg']['ends'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0032ca03-abbd-4ba2-b9c4-a08764510863",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/cerberus/ad003/5xCLU-h2kbKI_HO_F_4_months_HC_1_ic.tsv']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "prev_wc = {'cerberus_run': str(int(wc['cerberus_run'])-1)}\n",
    "prev_wc\n",
    "\n",
    "get_cfg_entries({'cerberus_run': str(int(wc['cerberus_run'])-1)}, df, config['cerberus']['ics'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "fa9a8cd6-cf66-46f5-a465-259bf15ddf71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/cerberus/agg/ad005_BIN1_HO_F_4_months_HC_3_ic.tsv']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_cfg_entries({'cerberus_run':df.cerberus_run.max()},\n",
    "                        df,\n",
    "                        config['cerberus']['agg']['ics'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09862927-b2ee-4d40-8a7e-30a069f592d4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9be4fc03-dea2-4c32-87ce-3e520fb2c0b1",
   "metadata": {},
   "source": [
    "## 231019 pipeline restructuring -- cerberus getting source from current wcs (should be the study_genotype_sex_age_tissue combo again)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6a69e133-6344-4c37-9521-4ad9a8255dca",
   "metadata": {},
   "outputs": [],
   "source": [
    "wc = {'study': 'ad003',\n",
    "      'genotype': '5xFADHEMI',\n",
    "      'sex': 'F',\n",
    "      'age': '4_months',\n",
    "      'tissue': 'HC',\n",
    "      'end_mode': 'tss',\n",
    "      'cerberus_run': '2'}\n",
    "\n",
    "config_tsv = '231018_config.tsv'\n",
    "# config_tsv = 'test_local_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a63678d2-884e-4eb2-8a65-2d679ee40fcc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ad003_5xFADHEMI_F_4_months_HC'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_df_col(wc, df, 'source')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7948e879-0091-4743-b103-42dbacaec2d3",
   "metadata": {},
   "source": [
    "## 231019 pipeline restructuring -- cerberus stuff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "e7b9ab25-87f9-459f-ad9d-4aa0be3b8283",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_tsv = '231018_config.tsv'\n",
    "# config_tsv = 'test_local_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "2e688db1-b45d-47b2-8532-2aa2f93463cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# assign a cerberus run to each \"sample\" (study+genotype+sex+age+tissue)\n",
    "# but first sort on study and sample such that they will always be ordered in the same way\n",
    "# this should freeze our results\n",
    "gb_cols = ['study', 'genotype', 'sex', 'age', 'tissue']\n",
    "df = df.sort_values(by=gb_cols, ascending=True)\n",
    "temp = df.copy(deep=True)\n",
    "temp = temp[gb_cols].groupby(gb_cols).count().reset_index()\n",
    "temp['cerberus_run'] = [i+1 for i in range(len(temp.index))]\n",
    "df = df.merge(temp, how='left', on=gb_cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "19c0ea9f-0664-4fb9-b792-f4cd8559c016",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>study</th>\n",
       "      <th>genotype</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>tissue</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "      <th>cerberus_run</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12517_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_12517_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12517</td>\n",
       "      <td>12517</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xCLU-h2kbKI_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12517_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_12517_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12517</td>\n",
       "      <td>12517</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xCLU-h2kbKI_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC_1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12649_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_12649_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12649</td>\n",
       "      <td>12649</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xCLU-h2kbKI_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC_2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12649_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_12649_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12649</td>\n",
       "      <td>12649</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xCLU-h2kbKI_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC_2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12659_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_12659_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12659</td>\n",
       "      <td>12659</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xCLU-h2kbKI_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC_3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_12659_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_12659_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_12659</td>\n",
       "      <td>12659</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xCLU-h2kbKI_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xCLU-h2kbKI_HO_F_4_months_HC_3</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11616_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11616_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11617_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11617_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11617</td>\n",
       "      <td>11617</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11617_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11617_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11617</td>\n",
       "      <td>11617</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    batch platform                                              fname  \\\n",
       "0  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "1  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "2  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "3  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "4  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "5  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "6  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "7  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "8  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "9  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                         basename  \\\n",
       "0  ad003_12517_lig-blk_1_t1.fastq   \n",
       "1  ad003_12517_lig-blk_2_t1.fastq   \n",
       "2  ad003_12649_lig-blk_1_t1.fastq   \n",
       "3  ad003_12649_lig-blk_2_t1.fastq   \n",
       "4  ad003_12659_lig-blk_1_t1.fastq   \n",
       "5  ad003_12659_lig-blk_2_t1.fastq   \n",
       "6  ad003_11616_lig-blk_1_t1.fastq   \n",
       "7  ad003_11616_lig-blk_2_t1.fastq   \n",
       "8  ad003_11617_lig-blk_1_t1.fastq   \n",
       "9  ad003_11617_lig-blk_2_t1.fastq   \n",
       "\n",
       "                                                path flowcell  \\\n",
       "0  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "1  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "2  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "3  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "4  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "5  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "6  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "7  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "8  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "9  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "\n",
       "               file_stem  chop_num  sample_temp  mouse_id  study  \\\n",
       "0  ad003_12517_lig-blk_1       1.0  ad003_12517     12517  ad003   \n",
       "1  ad003_12517_lig-blk_2       1.0  ad003_12517     12517  ad003   \n",
       "2  ad003_12649_lig-blk_1       1.0  ad003_12649     12649  ad003   \n",
       "3  ad003_12649_lig-blk_2       1.0  ad003_12649     12649  ad003   \n",
       "4  ad003_12659_lig-blk_1       1.0  ad003_12659     12659  ad003   \n",
       "5  ad003_12659_lig-blk_2       1.0  ad003_12659     12659  ad003   \n",
       "6  ad003_11616_lig-blk_1       1.0  ad003_11616     11616  ad003   \n",
       "7  ad003_11616_lig-blk_2       1.0  ad003_11616     11616  ad003   \n",
       "8  ad003_11617_lig-blk_1       1.0  ad003_11617     11617  ad003   \n",
       "9  ad003_11617_lig-blk_2       1.0  ad003_11617     11617  ad003   \n",
       "\n",
       "          genotype sex       age tissue                         sample  \\\n",
       "0  5xCLU-h2kbKI_HO   F  4_months     HC  5xCLU-h2kbKI_HO_F_4_months_HC   \n",
       "1  5xCLU-h2kbKI_HO   F  4_months     HC  5xCLU-h2kbKI_HO_F_4_months_HC   \n",
       "2  5xCLU-h2kbKI_HO   F  4_months     HC  5xCLU-h2kbKI_HO_F_4_months_HC   \n",
       "3  5xCLU-h2kbKI_HO   F  4_months     HC  5xCLU-h2kbKI_HO_F_4_months_HC   \n",
       "4  5xCLU-h2kbKI_HO   F  4_months     HC  5xCLU-h2kbKI_HO_F_4_months_HC   \n",
       "5  5xCLU-h2kbKI_HO   F  4_months     HC  5xCLU-h2kbKI_HO_F_4_months_HC   \n",
       "6        5xFADHEMI   F  4_months     HC        5xFADHEMI_F_4_months_HC   \n",
       "7        5xFADHEMI   F  4_months     HC        5xFADHEMI_F_4_months_HC   \n",
       "8        5xFADHEMI   F  4_months     HC        5xFADHEMI_F_4_months_HC   \n",
       "9        5xFADHEMI   F  4_months     HC        5xFADHEMI_F_4_months_HC   \n",
       "\n",
       "  biorep_num                          dataset  cerberus_run  \n",
       "0          1  5xCLU-h2kbKI_HO_F_4_months_HC_1             1  \n",
       "1          1  5xCLU-h2kbKI_HO_F_4_months_HC_1             1  \n",
       "2          2  5xCLU-h2kbKI_HO_F_4_months_HC_2             1  \n",
       "3          2  5xCLU-h2kbKI_HO_F_4_months_HC_2             1  \n",
       "4          3  5xCLU-h2kbKI_HO_F_4_months_HC_3             1  \n",
       "5          3  5xCLU-h2kbKI_HO_F_4_months_HC_3             1  \n",
       "6          1        5xFADHEMI_F_4_months_HC_1             2  \n",
       "7          1        5xFADHEMI_F_4_months_HC_1             2  \n",
       "8          2        5xFADHEMI_F_4_months_HC_2             2  \n",
       "9          2        5xFADHEMI_F_4_months_HC_2             2  "
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ceedc3f-573b-4f77-b7b7-1b92970a3513",
   "metadata": {},
   "source": [
    "## 231019 pipeline restructuring -- lapa filtering settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "6dc0f06f-b75d-4213-983b-cfeae457efbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "t_novs = config['lapa']['filt']['t_novs']\n",
    "g_novs = config['lapa']['filt']['g_novs']\n",
    "filt_spikes = config['lapa']['filt']['remove_spikes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "aeacce71-5b11-4602-b2b6-0c5647af4da1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "bool"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(filt_spikes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81b30ede-48ba-4fe6-89ba-85e67869ab83",
   "metadata": {},
   "source": [
    "## 231019 pipeline restructuring -- lapa input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "5a084302-9de1-46d9-a5ca-6d4ab5d7399c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def subset_df_on_wcs(wc, df):\n",
    "    \"\"\"\n",
    "    Return a copy of the input metadata df limited to the wildcards\n",
    "    \"\"\"\n",
    "    temp = df.copy(deep=True)\n",
    "    for key, item in wc.items():\n",
    "        \n",
    "        # for entries that we don't have in the df, such as LAPA end mode\n",
    "        if key not in df.columns:\n",
    "            continue\n",
    "        \n",
    "        # if we're given a list of possibilities\n",
    "        if type(item) == list:\n",
    "            temp = temp.loc[temp[key].isin(item)]\n",
    "        else:\n",
    "            temp = temp.loc[temp[key] == item]\n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "95196391-8fc2-4932-b3f6-b323e7499023",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cfg_entries(wc, df, cfg_entry, return_df=False):\n",
    "    \"\"\"\n",
    "    Expand a config entry based on the wildcards and the\n",
    "    values in df that satisfy these wildcards\n",
    "\n",
    "    Parameters:\n",
    "        return_df (bool): Return DataFrame with 'file' column\n",
    "            as opposed to list of files. Default: False\n",
    "    \"\"\"\n",
    "    temp = subset_df_on_wcs(wc, df)\n",
    "\n",
    "    # cols = ['study', 'genotype', 'sex',\n",
    "    #         'age', 'tissue', 'biorep_num',\n",
    "    #         'flowcell']\n",
    "\n",
    "    study = temp.study.tolist()\n",
    "    genotype = temp.genotype.tolist()\n",
    "    sex = temp.sex.tolist()\n",
    "    age = temp.age.tolist()\n",
    "    tissue = temp.tissue.tolist()\n",
    "    biorep_num = temp.biorep_num.tolist()\n",
    "    flowcell = temp.flowcell.tolist()\n",
    "\n",
    "    files = expand(cfg_entry,\n",
    "                   zip,\n",
    "                   study=study,\n",
    "                   genotype=genotype,\n",
    "                   sex=sex,\n",
    "                   age=age,\n",
    "                   tissue=tissue,\n",
    "                   biorep_num=biorep_num,\n",
    "                   flowcell=flowcell,\n",
    "                   allow_missing=True)\n",
    "\n",
    "    temp['file'] = files\n",
    "\n",
    "    # make sure we only take unique ones\n",
    "    # cols.append('file')\n",
    "    # temp = temp[cols]\n",
    "    temp = temp.drop_duplicates(subset='file', keep='first')\n",
    "    files = temp['file'].tolist()\n",
    "\n",
    "    if return_df:\n",
    "        return temp\n",
    "    else:\n",
    "        return files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "ead35f3e-264b-4cf7-9489-81ef7764d968",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_lapa_settings(wc, cfg_entry, kind):\n",
    "    \"\"\"\n",
    "    Get the command name or file output name\n",
    "    given whether we're running LAPA in tss or tes mode\n",
    "\n",
    "    Parameters:\n",
    "        lapa_ends (str): Config entry \n",
    "        kind (str): {'temp_file', 'lapa_cmd'}\n",
    "    \"\"\"\n",
    "    \n",
    "    # get the output file from the lapa run\n",
    "    lapa_ends = get_cfg_entries(wc, df, cfg_entry, return_df=False)\n",
    "    \n",
    "    # now expand for end mode\n",
    "    lapa_ends = expand(lapa_ends, \n",
    "                       zip=True,\n",
    "                       end_mode=wc['end_mode'])    \n",
    "    assert len(list(set(lapa_ends))) == 1\n",
    "    lapa_ends = lapa_ends[0]\n",
    "    \n",
    "    if kind == 'temp_file':\n",
    "        temp = os.path.dirname(lapa_ends)+'/'\n",
    "        if wc['end_mode'] == 'tes':\n",
    "            temp += 'polyA_clusters.bed'\n",
    "        elif wc['end_mode'] == 'tss':\n",
    "            temp += 'tss_clusters.bed'\n",
    "        return temp\n",
    "    elif kind == 'lapa_cmd':\n",
    "        if wc['end_mode'] == 'tes':\n",
    "            return 'lapa'\n",
    "        elif wc['end_mode'] == 'tss':\n",
    "            return 'lapa_tss'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "824801c0-fc32-49b5-b9bc-86aa97db5fdd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'data/lapa/ad003/tss/tss_clusters.bed'"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get all files from bioreps of same genotype, sex, age, tissue\n",
    "wc = {'study': 'ad003',\n",
    "      'genotype': '5xFADHEMI',\n",
    "      'sex': 'F',\n",
    "      'age': ['4_months', '10_days'],\n",
    "      'tissue': 'HC',\n",
    "      'end_mode': 'tss'}\n",
    "\n",
    "cfg_entry = config['lapa']['ends']\n",
    "get_lapa_settings(wc, cfg_entry, 'temp_file')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "b2365cc7-6d39-4f12-9634-5b5eb787177d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'lapa_tss'"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_lapa_settings(wc, cfg_entry, 'lapa_cmd')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b510dd70-376c-42ee-9f09-5ca61be5f9e4",
   "metadata": {},
   "source": [
    "## 231019 pipeline restructuring -- talon input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "2cbbbb5e-3628-4627-9437-ecc516f99473",
   "metadata": {},
   "outputs": [],
   "source": [
    "# utilities related to defining snakemake rules\n",
    "from snakemake.io import expand\n",
    "\n",
    "def subset_df_on_wcs(wc, df):\n",
    "    \"\"\"\n",
    "    Return a copy of the input metadata df limited to the wildcards\n",
    "    \"\"\"\n",
    "    temp = df.copy(deep=True)\n",
    "    for key, item in wc.items():\n",
    "        if type(item) == list:\n",
    "            temp = temp.loc[temp[key].isin(item)]\n",
    "        else:\n",
    "            temp = temp.loc[temp[key] == item]\n",
    "    return temp\n",
    "\n",
    "def get_df_col(wc, df, col):\n",
    "    \"\"\"\n",
    "    From the metadata dataframe df, get the entries that satisfy\n",
    "    the wildcards requirements and return the corresponding value\n",
    "    from col. Ensure that this is always a 1:1 relationship, otherwise\n",
    "    throw an error.\n",
    "    \"\"\"\n",
    "    cols = [col] + [key for key, item in wc.items()]\n",
    "\n",
    "    temp = subset_df_on_wcs(wc, df)\n",
    "    temp = temp[cols].drop_duplicates()\n",
    "\n",
    "    if len(temp.index) != 1:\n",
    "        msg = 'Issues getting data from DF with wildcards'\n",
    "        for key, item in wc.items():\n",
    "            msg+=f'\\n{key}: {item}'\n",
    "        raise ValueError(msg)\n",
    "\n",
    "    val = temp[col].tolist()[0]\n",
    "    return val\n",
    "\n",
    "def get_cfg_entries(wc, df, cfg_entry, return_df=False):\n",
    "    \"\"\"\n",
    "    Expand a config entry based on the wildcards and the\n",
    "    values in df that satisfy these wildcards\n",
    "    \n",
    "    Parameters:\n",
    "        return_df (bool): Return DataFrame with 'file' column \n",
    "            as opposed to list of files. Default: False\n",
    "    \"\"\"\n",
    "    temp = subset_df_on_wcs(wc, df)\n",
    "\n",
    "    # cols = ['study', 'genotype', 'sex',\n",
    "    #         'age', 'tissue', 'biorep_num',\n",
    "    #         'flowcell']\n",
    "\n",
    "    study = temp.study.tolist()\n",
    "    genotype = temp.genotype.tolist()\n",
    "    sex = temp.sex.tolist()\n",
    "    age = temp.age.tolist()\n",
    "    tissue = temp.tissue.tolist()\n",
    "    biorep_num = temp.biorep_num.tolist()\n",
    "    flowcell = temp.flowcell.tolist()\n",
    "\n",
    "    files = expand(cfg_entry,\n",
    "                   zip,\n",
    "                   study=study,\n",
    "                   genotype=genotype,\n",
    "                   sex=sex,\n",
    "                   age=age,\n",
    "                   tissue=tissue,\n",
    "                   biorep_num=biorep_num,\n",
    "                   flowcell=flowcell)\n",
    "\n",
    "    temp['file'] = files\n",
    "\n",
    "    # make sure we only take unique ones\n",
    "    # cols.append('file')\n",
    "    # temp = temp[cols]\n",
    "    temp = temp.drop_duplicates(subset='file', keep='first')\n",
    "    files = temp['file'].tolist()\n",
    "    \n",
    "    if return_df:\n",
    "        return temp\n",
    "    else: \n",
    "        return files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "b456714c-7bd4-423e-9efa-01b3312ca885",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>study</th>\n",
       "      <th>genotype</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>tissue</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "      <th>file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11616_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "      <td>data/lapa/ad003/{end_mode}/5xFADHEMI_F_4_month...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    batch platform                                              fname  \\\n",
       "0  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                         basename  \\\n",
       "0  ad003_11616_lig-blk_1_t1.fastq   \n",
       "\n",
       "                                                path flowcell  \\\n",
       "0  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "\n",
       "               file_stem  chop_num  sample_temp  mouse_id  study   genotype  \\\n",
       "0  ad003_11616_lig-blk_1       1.0  ad003_11616     11616  ad003  5xFADHEMI   \n",
       "\n",
       "  sex       age tissue                   sample biorep_num  \\\n",
       "0   F  4_months     HC  5xFADHEMI_F_4_months_HC          1   \n",
       "\n",
       "                     dataset  \\\n",
       "0  5xFADHEMI_F_4_months_HC_1   \n",
       "\n",
       "                                                file  \n",
       "0  data/lapa/ad003/{end_mode}/5xFADHEMI_F_4_month...  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get all files from bioreps of same genotype, sex, age, tissue\n",
    "wc = {'study': 'ad003',\n",
    "      'genotype': '5xFADHEMI',\n",
    "      'sex': 'F',\n",
    "      'age': ['4_months', '10_days'],\n",
    "      'tissue': 'HC'}\n",
    "\n",
    "files = get_cfg_entries(wc, df, config['merge']['sort_bam'])\n",
    "files\n",
    "temp = get_cfg_entries(wc, df, cfg_entry, return_df=True)\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "406687a6-c2f4-4255-bdd4-596f2a660803",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>study</th>\n",
       "      <th>genotype</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>tissue</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11616_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11616_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11617_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11617_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11617</td>\n",
       "      <td>11617</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11617_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11617_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11617</td>\n",
       "      <td>11617</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11625_lig-blk_1_t2.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11625_lig-blk_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ad003_11625</td>\n",
       "      <td>11625</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11625_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11625_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11625</td>\n",
       "      <td>11625</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    batch platform                                              fname  \\\n",
       "0  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "1  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "2  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "3  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "4  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "5  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                         basename  \\\n",
       "0  ad003_11616_lig-blk_1_t1.fastq   \n",
       "1  ad003_11616_lig-blk_2_t1.fastq   \n",
       "2  ad003_11617_lig-blk_1_t1.fastq   \n",
       "3  ad003_11617_lig-blk_2_t1.fastq   \n",
       "4  ad003_11625_lig-blk_1_t2.fastq   \n",
       "5  ad003_11625_lig-blk_2_t1.fastq   \n",
       "\n",
       "                                                path flowcell  \\\n",
       "0  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "1  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "2  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "3  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "4  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "5  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "\n",
       "               file_stem  chop_num  sample_temp  mouse_id  study   genotype  \\\n",
       "0  ad003_11616_lig-blk_1       1.0  ad003_11616     11616  ad003  5xFADHEMI   \n",
       "1  ad003_11616_lig-blk_2       1.0  ad003_11616     11616  ad003  5xFADHEMI   \n",
       "2  ad003_11617_lig-blk_1       1.0  ad003_11617     11617  ad003  5xFADHEMI   \n",
       "3  ad003_11617_lig-blk_2       1.0  ad003_11617     11617  ad003  5xFADHEMI   \n",
       "4  ad003_11625_lig-blk_1       2.0  ad003_11625     11625  ad003  5xFADHEMI   \n",
       "5  ad003_11625_lig-blk_2       1.0  ad003_11625     11625  ad003  5xFADHEMI   \n",
       "\n",
       "  sex       age tissue                   sample biorep_num  \\\n",
       "0   F  4_months     HC  5xFADHEMI_F_4_months_HC          1   \n",
       "1   F  4_months     HC  5xFADHEMI_F_4_months_HC          1   \n",
       "2   F  4_months     HC  5xFADHEMI_F_4_months_HC          2   \n",
       "3   F  4_months     HC  5xFADHEMI_F_4_months_HC          2   \n",
       "4   F  4_months     HC  5xFADHEMI_F_4_months_HC          3   \n",
       "5   F  4_months     HC  5xFADHEMI_F_4_months_HC          3   \n",
       "\n",
       "                     dataset  \n",
       "0  5xFADHEMI_F_4_months_HC_1  \n",
       "1  5xFADHEMI_F_4_months_HC_1  \n",
       "2  5xFADHEMI_F_4_months_HC_2  \n",
       "3  5xFADHEMI_F_4_months_HC_2  \n",
       "4  5xFADHEMI_F_4_months_HC_3  \n",
       "5  5xFADHEMI_F_4_months_HC_3  "
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = subset_df_on_wcs(wc, df)\n",
    "temp\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "6cc8d9a4-46e8-463b-b38f-d8093c1eda90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>study</th>\n",
       "      <th>genotype</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>tissue</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11616_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11616_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11617_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11617_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11617</td>\n",
       "      <td>11617</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11617_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11617_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11617</td>\n",
       "      <td>11617</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11625_lig-blk_1_t2.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11625_lig-blk_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ad003_11625</td>\n",
       "      <td>11625</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11625_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11625_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11625</td>\n",
       "      <td>11625</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    batch platform                                              fname  \\\n",
       "0  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "1  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "2  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "3  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "4  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "5  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                         basename  \\\n",
       "0  ad003_11616_lig-blk_1_t1.fastq   \n",
       "1  ad003_11616_lig-blk_2_t1.fastq   \n",
       "2  ad003_11617_lig-blk_1_t1.fastq   \n",
       "3  ad003_11617_lig-blk_2_t1.fastq   \n",
       "4  ad003_11625_lig-blk_1_t2.fastq   \n",
       "5  ad003_11625_lig-blk_2_t1.fastq   \n",
       "\n",
       "                                                path flowcell  \\\n",
       "0  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "1  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "2  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "3  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "4  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "5  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "\n",
       "               file_stem  chop_num  sample_temp  mouse_id  study   genotype  \\\n",
       "0  ad003_11616_lig-blk_1       1.0  ad003_11616     11616  ad003  5xFADHEMI   \n",
       "1  ad003_11616_lig-blk_2       1.0  ad003_11616     11616  ad003  5xFADHEMI   \n",
       "2  ad003_11617_lig-blk_1       1.0  ad003_11617     11617  ad003  5xFADHEMI   \n",
       "3  ad003_11617_lig-blk_2       1.0  ad003_11617     11617  ad003  5xFADHEMI   \n",
       "4  ad003_11625_lig-blk_1       2.0  ad003_11625     11625  ad003  5xFADHEMI   \n",
       "5  ad003_11625_lig-blk_2       1.0  ad003_11625     11625  ad003  5xFADHEMI   \n",
       "\n",
       "  sex       age tissue                   sample biorep_num  \\\n",
       "0   F  4_months     HC  5xFADHEMI_F_4_months_HC          1   \n",
       "1   F  4_months     HC  5xFADHEMI_F_4_months_HC          1   \n",
       "2   F  4_months     HC  5xFADHEMI_F_4_months_HC          2   \n",
       "3   F  4_months     HC  5xFADHEMI_F_4_months_HC          2   \n",
       "4   F  4_months     HC  5xFADHEMI_F_4_months_HC          3   \n",
       "5   F  4_months     HC  5xFADHEMI_F_4_months_HC          3   \n",
       "\n",
       "                     dataset  \n",
       "0  5xFADHEMI_F_4_months_HC_1  \n",
       "1  5xFADHEMI_F_4_months_HC_1  \n",
       "2  5xFADHEMI_F_4_months_HC_2  \n",
       "3  5xFADHEMI_F_4_months_HC_2  \n",
       "4  5xFADHEMI_F_4_months_HC_3  \n",
       "5  5xFADHEMI_F_4_months_HC_3  "
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df.loc[(df.study==wc['study'])&\\\n",
    "       (df.genotype==wc['genotype'])&\\\n",
    "       (df.sex==wc['sex'])&\\\n",
    "       (df.age.isin(wc['age']))&\\\n",
    "       (df.tissue==wc['tissue'])]\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "83b3c9e7-2637-4d18-b1f1-6af3778ebab3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>dataset</th>\n",
       "      <th>platform</th>\n",
       "      <th>file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "      <td>ONT</td>\n",
       "      <td>data/flowcell_merge/5xFADHEMI_F_4_months_HC_1.bam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "      <td>ONT</td>\n",
       "      <td>data/flowcell_merge/5xFADHEMI_F_4_months_HC_2.bam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_3</td>\n",
       "      <td>ONT</td>\n",
       "      <td>data/flowcell_merge/5xFADHEMI_F_4_months_HC_3.bam</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    sample                    dataset platform  \\\n",
       "0  5xFADHEMI_F_4_months_HC  5xFADHEMI_F_4_months_HC_1      ONT   \n",
       "2  5xFADHEMI_F_4_months_HC  5xFADHEMI_F_4_months_HC_2      ONT   \n",
       "4  5xFADHEMI_F_4_months_HC  5xFADHEMI_F_4_months_HC_3      ONT   \n",
       "\n",
       "                                                file  \n",
       "0  data/flowcell_merge/5xFADHEMI_F_4_months_HC_1.bam  \n",
       "2  data/flowcell_merge/5xFADHEMI_F_4_months_HC_2.bam  \n",
       "4  data/flowcell_merge/5xFADHEMI_F_4_months_HC_3.bam  "
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp[['sample', 'dataset', 'platform', 'file']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "9cc288eb-79a1-4664-b7f2-610b5f2afbba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>study</th>\n",
       "      <th>genotype</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>tissue</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "      <th>file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11616_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "      <td>data/flowcell_merge/5xFADHEMI_F_4_months_HC_1.bam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11617_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11617_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11617</td>\n",
       "      <td>11617</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "      <td>data/flowcell_merge/5xFADHEMI_F_4_months_HC_2.bam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11625_lig-blk_1_t2.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11625_lig-blk_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ad003_11625</td>\n",
       "      <td>11625</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_3</td>\n",
       "      <td>data/flowcell_merge/5xFADHEMI_F_4_months_HC_3.bam</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    batch platform                                              fname  \\\n",
       "0  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "2  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "4  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                         basename  \\\n",
       "0  ad003_11616_lig-blk_1_t1.fastq   \n",
       "2  ad003_11617_lig-blk_1_t1.fastq   \n",
       "4  ad003_11625_lig-blk_1_t2.fastq   \n",
       "\n",
       "                                                path flowcell  \\\n",
       "0  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "2  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "4  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "\n",
       "               file_stem  chop_num  sample_temp  mouse_id  study   genotype  \\\n",
       "0  ad003_11616_lig-blk_1       1.0  ad003_11616     11616  ad003  5xFADHEMI   \n",
       "2  ad003_11617_lig-blk_1       1.0  ad003_11617     11617  ad003  5xFADHEMI   \n",
       "4  ad003_11625_lig-blk_1       2.0  ad003_11625     11625  ad003  5xFADHEMI   \n",
       "\n",
       "  sex       age tissue                   sample biorep_num  \\\n",
       "0   F  4_months     HC  5xFADHEMI_F_4_months_HC          1   \n",
       "2   F  4_months     HC  5xFADHEMI_F_4_months_HC          2   \n",
       "4   F  4_months     HC  5xFADHEMI_F_4_months_HC          3   \n",
       "\n",
       "                     dataset  \\\n",
       "0  5xFADHEMI_F_4_months_HC_1   \n",
       "2  5xFADHEMI_F_4_months_HC_2   \n",
       "4  5xFADHEMI_F_4_months_HC_3   \n",
       "\n",
       "                                                file  \n",
       "0  data/flowcell_merge/5xFADHEMI_F_4_months_HC_1.bam  \n",
       "2  data/flowcell_merge/5xFADHEMI_F_4_months_HC_2.bam  \n",
       "4  data/flowcell_merge/5xFADHEMI_F_4_months_HC_3.bam  "
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "83a767d5-7b36-426b-805a-f453e43e7d2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>study</th>\n",
       "      <th>genotype</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>tissue</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad005_13051_lig-blk_2_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad005_13051_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad005_13051</td>\n",
       "      <td>13051</td>\n",
       "      <td>ad005</td>\n",
       "      <td>BIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>BIN1_HO_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>BIN1_HO_F_4_months_HC_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad005_13036_lig-blk_2_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad005_13036_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad005_13036</td>\n",
       "      <td>13036</td>\n",
       "      <td>ad005</td>\n",
       "      <td>BIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>BIN1_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>BIN1_HO_F_4_months_HC_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad005_13047_lig-blk_3_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad005_13047_lig-blk_3</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad005_13047</td>\n",
       "      <td>13047</td>\n",
       "      <td>ad005</td>\n",
       "      <td>BIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>BIN1_HO_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>BIN1_HO_F_4_months_HC_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad005_13051_lig-blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad005_13051_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad005_13051</td>\n",
       "      <td>13051</td>\n",
       "      <td>ad005</td>\n",
       "      <td>BIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>BIN1_HO_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>BIN1_HO_F_4_months_HC_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad005_13053_lig-blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad005_13053_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad005_13053</td>\n",
       "      <td>13053</td>\n",
       "      <td>ad005</td>\n",
       "      <td>5xBIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xBIN1_HO_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xBIN1_HO_F_4_months_HC_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad005_13022_lig-blk_2_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad005_13022_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad005_13022</td>\n",
       "      <td>13022</td>\n",
       "      <td>ad005</td>\n",
       "      <td>5xBIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xBIN1_HO_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xBIN1_HO_F_4_months_HC_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad005_13053_lig-blk_2_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad005_13053_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad005_13053</td>\n",
       "      <td>13053</td>\n",
       "      <td>ad005</td>\n",
       "      <td>5xBIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xBIN1_HO_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xBIN1_HO_F_4_months_HC_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad005_13019_lig-blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad005_13019_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad005_13019</td>\n",
       "      <td>13019</td>\n",
       "      <td>ad005</td>\n",
       "      <td>5xBIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xBIN1_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xBIN1_HO_F_4_months_HC_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad005_13047_lig-blk_2_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad005_13047_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad005_13047</td>\n",
       "      <td>13047</td>\n",
       "      <td>ad005</td>\n",
       "      <td>BIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>BIN1_HO_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>BIN1_HO_F_4_months_HC_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad005_13022_lig-blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad005_13022_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad005_13022</td>\n",
       "      <td>13022</td>\n",
       "      <td>ad005</td>\n",
       "      <td>5xBIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xBIN1_HO_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xBIN1_HO_F_4_months_HC_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad005_13019_lig-blk_2_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad005_13019_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad005_13019</td>\n",
       "      <td>13019</td>\n",
       "      <td>ad005</td>\n",
       "      <td>5xBIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xBIN1_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xBIN1_HO_F_4_months_HC_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad005_13047_lig-blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>3</td>\n",
       "      <td>ad005_13047_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad005_13047</td>\n",
       "      <td>13047</td>\n",
       "      <td>ad005</td>\n",
       "      <td>BIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>BIN1_HO_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>BIN1_HO_F_4_months_HC_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad005_13036_lig-blk_1_t1.fastq.gz</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad005_13036_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad005_13036</td>\n",
       "      <td>13036</td>\n",
       "      <td>ad005</td>\n",
       "      <td>BIN1_HO</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>BIN1_HO_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>BIN1_HO_F_4_months_HC_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     batch platform                                              fname  \\\n",
       "24  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "25  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "26  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "27  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "28  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "29  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "30  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "31  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "32  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "33  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "34  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "35  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "36  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                             basename  \\\n",
       "24  ad005_13051_lig-blk_2_t1.fastq.gz   \n",
       "25  ad005_13036_lig-blk_2_t1.fastq.gz   \n",
       "26  ad005_13047_lig-blk_3_t1.fastq.gz   \n",
       "27  ad005_13051_lig-blk_1_t1.fastq.gz   \n",
       "28  ad005_13053_lig-blk_1_t1.fastq.gz   \n",
       "29  ad005_13022_lig-blk_2_t1.fastq.gz   \n",
       "30  ad005_13053_lig-blk_2_t1.fastq.gz   \n",
       "31  ad005_13019_lig-blk_1_t1.fastq.gz   \n",
       "32  ad005_13047_lig-blk_2_t1.fastq.gz   \n",
       "33  ad005_13022_lig-blk_1_t1.fastq.gz   \n",
       "34  ad005_13019_lig-blk_2_t1.fastq.gz   \n",
       "35  ad005_13047_lig-blk_1_t1.fastq.gz   \n",
       "36  ad005_13036_lig-blk_1_t1.fastq.gz   \n",
       "\n",
       "                                                 path flowcell  \\\n",
       "24  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "25  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "26  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "27  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "28  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "29  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "30  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "31  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "32  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "33  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "34  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "35  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        3   \n",
       "36  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "\n",
       "                file_stem  chop_num  sample_temp  mouse_id  study   genotype  \\\n",
       "24  ad005_13051_lig-blk_2       1.0  ad005_13051     13051  ad005    BIN1_HO   \n",
       "25  ad005_13036_lig-blk_2       1.0  ad005_13036     13036  ad005    BIN1_HO   \n",
       "26  ad005_13047_lig-blk_3       1.0  ad005_13047     13047  ad005    BIN1_HO   \n",
       "27  ad005_13051_lig-blk_1       1.0  ad005_13051     13051  ad005    BIN1_HO   \n",
       "28  ad005_13053_lig-blk_1       1.0  ad005_13053     13053  ad005  5xBIN1_HO   \n",
       "29  ad005_13022_lig-blk_2       1.0  ad005_13022     13022  ad005  5xBIN1_HO   \n",
       "30  ad005_13053_lig-blk_2       1.0  ad005_13053     13053  ad005  5xBIN1_HO   \n",
       "31  ad005_13019_lig-blk_1       1.0  ad005_13019     13019  ad005  5xBIN1_HO   \n",
       "32  ad005_13047_lig-blk_2       1.0  ad005_13047     13047  ad005    BIN1_HO   \n",
       "33  ad005_13022_lig-blk_1       1.0  ad005_13022     13022  ad005  5xBIN1_HO   \n",
       "34  ad005_13019_lig-blk_2       1.0  ad005_13019     13019  ad005  5xBIN1_HO   \n",
       "35  ad005_13047_lig-blk_1       1.0  ad005_13047     13047  ad005    BIN1_HO   \n",
       "36  ad005_13036_lig-blk_1       1.0  ad005_13036     13036  ad005    BIN1_HO   \n",
       "\n",
       "   sex       age tissue                   sample biorep_num  \\\n",
       "24   F  4_months     HC    BIN1_HO_F_4_months_HC          3   \n",
       "25   F  4_months     HC    BIN1_HO_F_4_months_HC          1   \n",
       "26   F  4_months     HC    BIN1_HO_F_4_months_HC          2   \n",
       "27   F  4_months     HC    BIN1_HO_F_4_months_HC          3   \n",
       "28   F  4_months     HC  5xBIN1_HO_F_4_months_HC          3   \n",
       "29   F  4_months     HC  5xBIN1_HO_F_4_months_HC          2   \n",
       "30   F  4_months     HC  5xBIN1_HO_F_4_months_HC          3   \n",
       "31   F  4_months     HC  5xBIN1_HO_F_4_months_HC          1   \n",
       "32   F  4_months     HC    BIN1_HO_F_4_months_HC          2   \n",
       "33   F  4_months     HC  5xBIN1_HO_F_4_months_HC          2   \n",
       "34   F  4_months     HC  5xBIN1_HO_F_4_months_HC          1   \n",
       "35   F  4_months     HC    BIN1_HO_F_4_months_HC          2   \n",
       "36   F  4_months     HC    BIN1_HO_F_4_months_HC          1   \n",
       "\n",
       "                      dataset  \n",
       "24    BIN1_HO_F_4_months_HC_3  \n",
       "25    BIN1_HO_F_4_months_HC_1  \n",
       "26    BIN1_HO_F_4_months_HC_2  \n",
       "27    BIN1_HO_F_4_months_HC_3  \n",
       "28  5xBIN1_HO_F_4_months_HC_3  \n",
       "29  5xBIN1_HO_F_4_months_HC_2  \n",
       "30  5xBIN1_HO_F_4_months_HC_3  \n",
       "31  5xBIN1_HO_F_4_months_HC_1  \n",
       "32    BIN1_HO_F_4_months_HC_2  \n",
       "33  5xBIN1_HO_F_4_months_HC_2  \n",
       "34  5xBIN1_HO_F_4_months_HC_1  \n",
       "35    BIN1_HO_F_4_months_HC_2  \n",
       "36    BIN1_HO_F_4_months_HC_1  "
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.study=='ad005']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1c5d1b9-1641-4771-be8d-d190f363ec8f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d722e87-7836-45d4-9cbd-04141ab6796c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c4877da-71c2-4801-8d45-5631f3fb241a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79a456b8-a9c3-496f-a164-ce5cc5b5d483",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "452a8a4a-cdc9-48ad-bbbb-cc65f207438c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cc360584-28e8-4de8-958c-69ca9510c24b",
   "metadata": {},
   "source": [
    "## 231017 pipeline restructuring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e2d79393-02ee-4413-a8e0-5d8fd873bebf",
   "metadata": {},
   "outputs": [],
   "source": [
    "configfile: 'config.yml'\n",
    "config_tsv = '231018_config.tsv'\n",
    "meta_tsv = 'mouse_metadata.tsv'\n",
    "auto_dedupe = True\n",
    "df = parse_config_file(config_tsv,\n",
    "                       meta_tsv,\n",
    "                       auto_dedupe=auto_dedupe)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d94e3529-64e2-4660-a01d-6dcbf019e08d",
   "metadata": {},
   "outputs": [],
   "source": [
    "wc = {'genotype': '5xFADHEMI',\n",
    "      'sex': 'F',\n",
    "      'age': ['4_months', '10_days'],\n",
    "      'tissue': 'HC',\n",
    "      'biorep_num': '1',\n",
    "      'flowcell': '1'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "d66fda86-8a38-469a-a5eb-a10f37ceff0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data/talon_label/5xFADHEMI_F_4_months_HC_1_1_labeled_sorted.bam',\n",
       " 'data/talon_label/5xFADHEMI_F_4_months_HC_1_2_labeled_sorted.bam']"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cfg_entry = config['talon_label']['sort_bam']\n",
    "wc = {'genotype': '5xFADHEMI',\n",
    "      'sex': 'F',\n",
    "      'age': ['4_months', '10_days'],\n",
    "      'tissue': 'HC',\n",
    "      'biorep_num': '1'}\n",
    "\n",
    "temp = subset_df_on_wcs(wc, df)\n",
    "study = temp.study.tolist()\n",
    "genotype = temp.genotype.tolist()\n",
    "sex = temp.sex.tolist()\n",
    "age = temp.age.tolist()\n",
    "tissue = temp.tissue.tolist()\n",
    "biorep_num = temp.biorep_num.tolist()\n",
    "flowcell = temp.flowcell.tolist()\n",
    "\n",
    "files = expand(cfg_entry,\n",
    "               zip,\n",
    "               study=study,\n",
    "               genotype=genotype,\n",
    "               sex=sex,\n",
    "               age=age,\n",
    "               tissue=tissue,\n",
    "               biorep_num=biorep_num,\n",
    "               flowcell=flowcell)\n",
    "return files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "40bee402-6f45-4988-9b1c-7e27aadea48f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>batch</th>\n",
       "      <th>platform</th>\n",
       "      <th>fname</th>\n",
       "      <th>basename</th>\n",
       "      <th>path</th>\n",
       "      <th>flowcell</th>\n",
       "      <th>file_stem</th>\n",
       "      <th>chop_num</th>\n",
       "      <th>sample_temp</th>\n",
       "      <th>mouse_id</th>\n",
       "      <th>study</th>\n",
       "      <th>genotype</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>tissue</th>\n",
       "      <th>sample</th>\n",
       "      <th>biorep_num</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11616_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11616_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11616_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11616</td>\n",
       "      <td>11616</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>1</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11617_lig-blk_1_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11617_lig-blk_1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11617</td>\n",
       "      <td>11617</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11617_lig-blk_2_t1.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>2</td>\n",
       "      <td>ad003_11617_lig-blk_2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ad003_11617</td>\n",
       "      <td>11617</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>2</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>230516</td>\n",
       "      <td>ONT</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>ad003_11625_lig-blk_1_t2.fastq</td>\n",
       "      <td>/share/crsp/lab/seyedam/share/ad_nanopore_tmp/...</td>\n",
       "      <td>1</td>\n",
       "      <td>ad003_11625_lig-blk_1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ad003_11625</td>\n",
       "      <td>11625</td>\n",
       "      <td>ad003</td>\n",
       "      <td>5xFADHEMI</td>\n",
       "      <td>F</td>\n",
       "      <td>4_months</td>\n",
       "      <td>HC</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC</td>\n",
       "      <td>3</td>\n",
       "      <td>5xFADHEMI_F_4_months_HC_3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    batch platform                                              fname  \\\n",
       "0  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "1  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "2  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "3  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "4  230516      ONT  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...   \n",
       "\n",
       "                         basename  \\\n",
       "0  ad003_11616_lig-blk_1_t1.fastq   \n",
       "1  ad003_11616_lig-blk_2_t1.fastq   \n",
       "2  ad003_11617_lig-blk_1_t1.fastq   \n",
       "3  ad003_11617_lig-blk_2_t1.fastq   \n",
       "4  ad003_11625_lig-blk_1_t2.fastq   \n",
       "\n",
       "                                                path flowcell  \\\n",
       "0  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "1  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "2  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "3  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        2   \n",
       "4  /share/crsp/lab/seyedam/share/ad_nanopore_tmp/...        1   \n",
       "\n",
       "               file_stem  chop_num  sample_temp  mouse_id  study   genotype  \\\n",
       "0  ad003_11616_lig-blk_1       1.0  ad003_11616     11616  ad003  5xFADHEMI   \n",
       "1  ad003_11616_lig-blk_2       1.0  ad003_11616     11616  ad003  5xFADHEMI   \n",
       "2  ad003_11617_lig-blk_1       1.0  ad003_11617     11617  ad003  5xFADHEMI   \n",
       "3  ad003_11617_lig-blk_2       1.0  ad003_11617     11617  ad003  5xFADHEMI   \n",
       "4  ad003_11625_lig-blk_1       2.0  ad003_11625     11625  ad003  5xFADHEMI   \n",
       "\n",
       "  sex       age tissue                   sample biorep_num  \\\n",
       "0   F  4_months     HC  5xFADHEMI_F_4_months_HC          1   \n",
       "1   F  4_months     HC  5xFADHEMI_F_4_months_HC          1   \n",
       "2   F  4_months     HC  5xFADHEMI_F_4_months_HC          2   \n",
       "3   F  4_months     HC  5xFADHEMI_F_4_months_HC          2   \n",
       "4   F  4_months     HC  5xFADHEMI_F_4_months_HC          3   \n",
       "\n",
       "                     dataset  \n",
       "0  5xFADHEMI_F_4_months_HC_1  \n",
       "1  5xFADHEMI_F_4_months_HC_1  \n",
       "2  5xFADHEMI_F_4_months_HC_2  \n",
       "3  5xFADHEMI_F_4_months_HC_2  \n",
       "4  5xFADHEMI_F_4_months_HC_3  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "911dbea0-984a-4ed9-8d54-a8e7945e470c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/share/crsp/lab/seyedam/share/ad_nanopore_tmp/AD003/ad003_trimfastq/ad003_11616_lig-blk_1_t1.fastq'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_df_col(wc, df, 'fname')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3ef13cc8-043e-4ecb-baab-9ed4468af31a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replicates in modelad consist of \n",
    "# * genotype\n",
    "# * sex\n",
    "# * age\n",
    "# * tissue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95a1c223-7da3-41b1-9a2f-9741ef7f02e0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
